<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="SomaScan.db">
<title>Advanced Exploration of the SomaScan Menu • SomaScan.db</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Roboto-0.4.8/font.css" rel="stylesheet">
<link href="../deps/Source_Code_Pro-0.4.8/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Advanced Exploration of the SomaScan Menu">
<meta property="og:description" content="SomaScan.db">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#container" class="visually-hidden-focusable">Skip to content</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-none"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">SomaScan.db</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.99.8</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/SomaScan-db.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/SomaScan-db.html">Introduction to SomaScan.db</a>
    <a class="dropdown-item" href="../articles/example_adat_workflow.html">Example ADAT Workflow Using SomaScan.db</a>
    <a class="dropdown-item" href="../articles/advanced_usage_examples.html">Advanced Exploration of the SomaScan Menu</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/SomaLogic/SomaScan.db/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article" id="container">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Advanced Exploration of the SomaScan Menu</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/SomaLogic/SomaScan.db/blob/HEAD/vignettes/advanced_usage_examples.Rmd" class="external-link"><code>vignettes/advanced_usage_examples.Rmd</code></a></small>
      <div class="d-none name"><code>advanced_usage_examples.Rmd</code></div>
    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      Advanced use cases of SomaScan.db, including SomaScan menu
      exploration examples.
    </div>
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This vignette is a follow up to the “Introduction to SomaScan.db”
vignette, and will introduce more advanced capabilities of the
<code>SomaScan.db</code> package. Below we illustrate how
<code>SomaScan.db</code> can be used to deeply explore the SomaScan menu
and execute complex annotation functions, outside of the basic use of
<code>select</code> outlined in the introductory vignette. Knowledge of
SQL is not required, but a familiarity with R and SomaScan data is
highly suggested. For an introduction to the <code>SomaScan.db</code>
package and its methods, please see
<code><a href="../articles/SomaScan-db.html">vignette("SomaScan-db", "SomaScan.db")</a></code>.</p>
<p>Please note that this vignette will require the installation and
usage of <em>three</em> additional Bioconductor R packages: <em><a href="https://bioconductor.org/packages/3.18/GO.db" class="external-link">GO.db</a></em>,
<em><a href="https://bioconductor.org/packages/3.18/EnsDb.Hsapiens.v75" class="external-link">EnsDb.Hsapiens.v75</a></em>,
and <em><a href="https://bioconductor.org/packages/3.18/KEGGREST" class="external-link">KEGGREST</a></em>.
Please see the linked pages to find installation instructions for these
packages.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">GO.db</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">KEGGREST</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://somalogic.com" class="external-link">SomaScan.db</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://withr.r-lib.org" class="external-link">withr</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="incorporating-additional-annotations">Incorporating Additional Annotations<a class="anchor" aria-label="anchor" href="#incorporating-additional-annotations"></a>
</h2>
<div class="section level3">
<h3 id="gene-ontology">Gene Ontology<a class="anchor" aria-label="anchor" href="#gene-ontology"></a>
</h3>
<p>The <code>SomaScan.db</code> package allows a user to retrieve Gene
Ontology (GO) identifiers associated with a particular SomaScan
<code>SeqId</code> (or set of <code>SeqIds</code>). However, the
available GO annotations in <code>SomaScan.db</code> are limited; only
the GO ID, evidence code, and ontology category are currently available.
This helps prevent the package from accumulating an overwhelming number
of annotation elements, but limits the ability to extract detailed GO
information.</p>
<p>To illustrate this limitation, below we will display the GO terms
associated with the gene “IL31”:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">il31_go</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/select-SomaDb-method.html">select</a></span><span class="op">(</span><span class="va">SomaScan.db</span>, keys <span class="op">=</span> <span class="st">"IL31"</span>, keytype <span class="op">=</span> <span class="st">"SYMBOL"</span>, </span>
<span>                  columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PROBEID"</span>, <span class="st">"GO"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'select()' returned 1:many mapping between keys and columns</span></span></code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">il31_go</span></span></code></pre></div>
<pre><code><span><span class="co">##   SYMBOL   PROBEID         GO EVIDENCE ONTOLOGY</span></span>
<span><span class="co">## 1   IL31 10455-196 GO:0002376      IEA       BP</span></span>
<span><span class="co">## 2   IL31 10455-196 GO:0005125      IBA       MF</span></span>
<span><span class="co">## 3   IL31 10455-196 GO:0005126      IBA       MF</span></span>
<span><span class="co">## 4   IL31 10455-196 GO:0005515      IPI       MF</span></span>
<span><span class="co">## 5   IL31 10455-196 GO:0005576      TAS       CC</span></span>
<span><span class="co">## 6   IL31 10455-196 GO:0005615      IBA       CC</span></span>
<span><span class="co">## 7   IL31 10455-196 GO:0005615      IDA       CC</span></span>
<span><span class="co">## 8   IL31 10455-196 GO:0007165      IEA       BP</span></span></code></pre>
<p>In this data frame, IL31 maps to one single <code>SeqId</code>
(“10455-196”), indicated by the “PROBEID” column. This
<code>SeqId</code> and gene are associated with seven unique GO IDs (in
the “GO” column). The GO knowledgebase is vast, however, and these
identifiers are not particularly informative for anyone who hasn’t
memorized their more descriptive term names. Additional details for each
ID would make this table more informative and interpretable. Luckily,
there are two options for retrieving such data:</p>
<ol style="list-style-type: decimal">
<li>Using an additional set of GO-specific methods to retrieve
additional information for each GO ID (<code>Term</code>,
<code>Ontology</code>, <code>Definition</code>, and
<code>Synonym</code>)</li>
<li>Linking results from <code>SomaScan.db</code> with another
Bioconductor tool, like the <em><a href="https://bioconductor.org/packages/3.18/GO.db" class="external-link">GO.db</a></em>
annotation package</li>
</ol>
<p>Each of these techniques have their own special utility. Below, we
will work through examples of how the techniques described above can be
used to link GO information with the annotations from
<code>SomaScan.db</code>.</p>
<hr>
<div class="section level4">
<h4 id="go-methods">GO Methods<a class="anchor" aria-label="anchor" href="#go-methods"></a>
</h4>
<p>The <code>Term</code>, <code>Ontology</code>,
<code>Definition</code>, and <code>Synonym</code> methods are
GO-specific methods imported from the <code>AnnotationDbi</code>
package. They are designed to retrieve a single piece of information,
indicated by the method name, that corresponds to a set of GO
identifiers (note: we will skip <code>Ontology</code> in this vignette,
as the GO Ontology is already retrievable with
<code>SomaScan.db</code>).</p>
<p>The <code>Term</code> method retrieves a character string defining
the role of the gene product that corresponds to provided GO ID(s). In
the example below, we will retrieve the GO terms for each of the GO IDs
in the <code>select</code> results generated previously:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/GOTerms-class.html" class="external-link">Term</a></span><span class="op">(</span><span class="va">il31_go</span><span class="op">$</span><span class="va">GO</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                  GO:0002376                  GO:0005125 </span></span>
<span><span class="co">##     "immune system process"         "cytokine activity" </span></span>
<span><span class="co">##                  GO:0005126                  GO:0005515 </span></span>
<span><span class="co">## "cytokine receptor binding"           "protein binding" </span></span>
<span><span class="co">##                  GO:0005576                  GO:0005615 </span></span>
<span><span class="co">##      "extracellular region"       "extracellular space" </span></span>
<span><span class="co">##                  GO:0005615                  GO:0007165 </span></span>
<span><span class="co">##       "extracellular space"       "signal transduction"</span></span></code></pre>
<p>The <code>Definition</code> method retrieves a more detailed and
extended definition of the ontology for the input GO IDs:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/GOTerms-class.html" class="external-link">Definition</a></span><span class="op">(</span><span class="va">il31_go</span><span class="op">$</span><span class="va">GO</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        GO:0002376 </span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           "Any process involved in the development or functioning of the immune system, an organismal system for calibrated responses to potential internal or invasive threats." </span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        GO:0005125 </span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             "The activity of a soluble extracellular gene product that interacts with a receptor to effect a change in the activity of the receptor to control the survival, growth, differentiation and effector function of tissues and cells." </span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        GO:0005126 </span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 "Binding to a cytokine receptor." </span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        GO:0005515 </span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           "Binding to a protein." </span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        GO:0005576 </span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          "The space external to the outermost structure of a cell. For cells without external protective or external encapsulating structures this refers to space outside of the plasma membrane. This term covers the host cell environment outside an intracellular parasite." </span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        GO:0005615 </span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        "That part of a multicellular organism outside the cells proper, usually taken to be outside the plasma membranes, and occupied by fluid." </span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        GO:0005615 </span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        "That part of a multicellular organism outside the cells proper, usually taken to be outside the plasma membranes, and occupied by fluid." </span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        GO:0007165 </span></span>
<span><span class="co">## "The cellular process in which a signal is conveyed to trigger a change in the activity or state of a cell. Signal transduction begins with reception of a signal (e.g. a ligand binding to a receptor or receptor activation by a stimulus such as light), or for signal transduction in the absence of ligand, signal-withdrawal or the activity of a constitutively active receptor. Signal transduction ends with regulation of a downstream cellular process, e.g. regulation of transcription or regulation of a metabolic process. Signal transduction covers signaling from receptors located on the surface of the cell and signaling via molecules located within the cell. For signaling between cells, signal transduction is restricted to events at and within the receiving cell."</span></span></code></pre>
<p>And finally, the <code>Synonym</code> method can be used to retrieve
other ontology terms that are considered to be synonymous to the primary
term attached to the GO ID. For example, “type I programmed cell death”
is considered synonymous with “apoptosis”. It’s worth noting that
<code>Synonym</code> can return a large set of results, so we caution
against providing a large set of GO IDs to <code>Synonym</code>:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/GOTerms-class.html" class="external-link">Synonym</a></span><span class="op">(</span><span class="va">il31_go</span><span class="op">$</span><span class="va">GO</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $&lt;NA&gt;</span></span>
<span><span class="co">## NULL</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $`GO:0005125`</span></span>
<span><span class="co">## [1] "autocrine activity" "paracrine activity"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $`GO:0005126`</span></span>
<span><span class="co">## [1] "hematopoietin/interferon-class (D200-domain) cytokine receptor binding"</span></span>
<span><span class="co">## [2] "hematopoietin/interferon-class (D200-domain) cytokine receptor ligand" </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $`GO:0005515`</span></span>
<span><span class="co">## [1] "GO:0001948"                 "GO:0045308"                </span></span>
<span><span class="co">## [3] "protein amino acid binding" "glycoprotein binding"      </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $`GO:0005576`</span></span>
<span><span class="co">## [1] "extracellular"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $`GO:0005615`</span></span>
<span><span class="co">## [1] "intercellular space"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $`GO:0005615`</span></span>
<span><span class="co">## [1] "intercellular space"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $`GO:0007165`</span></span>
<span><span class="co">##  [1] "GO:0023014"                                      </span></span>
<span><span class="co">##  [2] "GO:0023015"                                      </span></span>
<span><span class="co">##  [3] "GO:0023016"                                      </span></span>
<span><span class="co">##  [4] "GO:0023033"                                      </span></span>
<span><span class="co">##  [5] "GO:0023045"                                      </span></span>
<span><span class="co">##  [6] "signaling pathway"                               </span></span>
<span><span class="co">##  [7] "signalling pathway"                              </span></span>
<span><span class="co">##  [8] "signal transduction by cis-phosphorylation"      </span></span>
<span><span class="co">##  [9] "signal transduction by conformational transition"</span></span>
<span><span class="co">## [10] "signal transduction by protein phosphorylation"  </span></span>
<span><span class="co">## [11] "signal transduction by trans-phosphorylation"    </span></span>
<span><span class="co">## [12] "signaling cascade"                               </span></span>
<span><span class="co">## [13] "signalling cascade"</span></span></code></pre>
<p>A GO synonym was not found for the first identifier in the provided
vector, so an <code>NA</code> was returned.</p>
<p>These functions are useful for quickly retrieving information for a
given GO ID, but you’ll notice that the results are returned as a vector
or list, rather than a data frame. Depending on the application, this
may be useful - for example, these methods are handy for on-the-fly GO
term or definition lookups, but their format can be cumbersome to
incorporate into a data frame created by <code>select</code>.</p>
<p>Let’s return to the <code>il31_go</code> data frame we generated
previously. How can we incorporate the additional information obtained
by <code>Term</code>, <code>Definition</code>, and <code>Synonym</code>
into this object? Assuming the output is the same length as the number
of rows in <code>il31_go</code>, the character vector obtained by
<code>Term</code>, <code>Definition</code>, or <code>Synonym</code>, can
be easily appended as a new column in the <code>il31_go</code> data
frame:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">trms</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/GOTerms-class.html" class="external-link">Term</a></span><span class="op">(</span><span class="va">il31_go</span><span class="op">$</span><span class="va">GO</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">trms</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "character"</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">trms</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">il31_go</span><span class="op">$</span><span class="va">GO</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">il31_go</span><span class="op">$</span><span class="va">TERM</span> <span class="op">&lt;-</span> <span class="va">trms</span></span>
<span><span class="va">il31_go</span></span></code></pre></div>
<pre><code><span><span class="co">##   SYMBOL   PROBEID         GO EVIDENCE ONTOLOGY                      TERM</span></span>
<span><span class="co">## 1   IL31 10455-196 GO:0002376      IEA       BP     immune system process</span></span>
<span><span class="co">## 2   IL31 10455-196 GO:0005125      IBA       MF         cytokine activity</span></span>
<span><span class="co">## 3   IL31 10455-196 GO:0005126      IBA       MF cytokine receptor binding</span></span>
<span><span class="co">## 4   IL31 10455-196 GO:0005515      IPI       MF           protein binding</span></span>
<span><span class="co">## 5   IL31 10455-196 GO:0005576      TAS       CC      extracellular region</span></span>
<span><span class="co">## 6   IL31 10455-196 GO:0005615      IBA       CC       extracellular space</span></span>
<span><span class="co">## 7   IL31 10455-196 GO:0005615      IDA       CC       extracellular space</span></span>
<span><span class="co">## 8   IL31 10455-196 GO:0007165      IEA       BP       signal transduction</span></span></code></pre>
<p>The same can be done with the output of <code>Definition</code>:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">defs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/GOTerms-class.html" class="external-link">Definition</a></span><span class="op">(</span><span class="va">il31_go</span><span class="op">$</span><span class="va">GO</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">defs</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "character"</span></span></code></pre>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">defs</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">il31_go</span><span class="op">$</span><span class="va">GO</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">il31_go</span><span class="op">$</span><span class="va">DEFINITION</span> <span class="op">&lt;-</span> <span class="va">defs</span></span>
<span><span class="va">il31_go</span><span class="op">[</span> ,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SYMBOL"</span>, <span class="st">"PROBEID"</span>, <span class="st">"GO"</span>, <span class="st">"TERM"</span>, <span class="st">"DEFINITION"</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##   SYMBOL   PROBEID         GO                      TERM</span></span>
<span><span class="co">## 1   IL31 10455-196 GO:0002376     immune system process</span></span>
<span><span class="co">## 2   IL31 10455-196 GO:0005125         cytokine activity</span></span>
<span><span class="co">## 3   IL31 10455-196 GO:0005126 cytokine receptor binding</span></span>
<span><span class="co">## 4   IL31 10455-196 GO:0005515           protein binding</span></span>
<span><span class="co">## 5   IL31 10455-196 GO:0005576      extracellular region</span></span>
<span><span class="co">## 6   IL31 10455-196 GO:0005615       extracellular space</span></span>
<span><span class="co">## 7   IL31 10455-196 GO:0005615       extracellular space</span></span>
<span><span class="co">## 8   IL31 10455-196 GO:0007165       signal transduction</span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        DEFINITION</span></span>
<span><span class="co">## 1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           Any process involved in the development or functioning of the immune system, an organismal system for calibrated responses to potential internal or invasive threats.</span></span>
<span><span class="co">## 2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             The activity of a soluble extracellular gene product that interacts with a receptor to effect a change in the activity of the receptor to control the survival, growth, differentiation and effector function of tissues and cells.</span></span>
<span><span class="co">## 3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 Binding to a cytokine receptor.</span></span>
<span><span class="co">## 4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           Binding to a protein.</span></span>
<span><span class="co">## 5                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          The space external to the outermost structure of a cell. For cells without external protective or external encapsulating structures this refers to space outside of the plasma membrane. This term covers the host cell environment outside an intracellular parasite.</span></span>
<span><span class="co">## 6                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        That part of a multicellular organism outside the cells proper, usually taken to be outside the plasma membranes, and occupied by fluid.</span></span>
<span><span class="co">## 7                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        That part of a multicellular organism outside the cells proper, usually taken to be outside the plasma membranes, and occupied by fluid.</span></span>
<span><span class="co">## 8 The cellular process in which a signal is conveyed to trigger a change in the activity or state of a cell. Signal transduction begins with reception of a signal (e.g. a ligand binding to a receptor or receptor activation by a stimulus such as light), or for signal transduction in the absence of ligand, signal-withdrawal or the activity of a constitutively active receptor. Signal transduction ends with regulation of a downstream cellular process, e.g. regulation of transcription or regulation of a metabolic process. Signal transduction covers signaling from receptors located on the surface of the cell and signaling via molecules located within the cell. For signaling between cells, signal transduction is restricted to events at and within the receiving cell.</span></span></code></pre>
<p>However, this only works cleanly when the output is a character
vector with the same order and number of elements as the input vector.
With the list output of <code>Synonym</code>, the process is a little
less straightforward. In addition, it takes multiple steps to generate
these additional annotations and combine them with a <code>select</code>
data frame. Instead of performing so many steps, we can utilize another
Bionconductor annotation resource called <em><a href="https://bioconductor.org/packages/3.18/GO.db" class="external-link">GO.db</a></em> to
retrieve GO annotation elements in a convenient data frame format.</p>
<hr>
</div>
<div class="section level4">
<h4 id="go-db">GO.db<a class="anchor" aria-label="anchor" href="#go-db"></a>
</h4>
<p>The <em><a href="https://bioconductor.org/packages/3.18/GO.db" class="external-link">GO.db</a></em> R
package contains annotations describing the entire Gene Ontology
knowledgebase, assembled using data directly from the <a href="http://geneontology.org/" class="external-link">GO website</a>. <em><a href="https://bioconductor.org/packages/3.18/GO.db" class="external-link">GO.db</a></em>
provides a method to easily retrieve the latest version of the Gene
Ontology knowledgebase into an R session. Like <code>SomaScan.db</code>,
<em><a href="https://bioconductor.org/packages/3.18/GO.db" class="external-link">GO.db</a></em> is an
annotation package that can be queried using the same five methods
(<code>select</code>, <code>keys</code>, <code>keytypes</code>,
<code>columns</code>, and <code>mapIds</code>). By utilizing both
<code>SomaScan.db</code> and <em><a href="https://bioconductor.org/packages/3.18/GO.db" class="external-link">GO.db</a></em>, it
is possible to connect <code>SeqIds</code> to GO IDs, then add
additional GO annotations that are not available within
<code>SomaScan.db</code>.</p>
<p>Let’s walk through an example. First, select a key (and corresponding
GO ID) to use as a starting point:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">go_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/select-SomaDb-method.html">select</a></span><span class="op">(</span><span class="va">SomaScan.db</span>, <span class="st">"IL3RA"</span>, keytype <span class="op">=</span> <span class="st">"SYMBOL"</span>,</span>
<span>                 columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GO"</span>, <span class="st">"SYMBOL"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'select()' returned 1:many mapping between keys and columns</span></span></code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">go_ids</span></span></code></pre></div>
<pre><code><span><span class="co">##    SYMBOL         GO EVIDENCE ONTOLOGY</span></span>
<span><span class="co">## 1   IL3RA GO:0004896      IBA       MF</span></span>
<span><span class="co">## 2   IL3RA GO:0004912      IDA       MF</span></span>
<span><span class="co">## 3   IL3RA GO:0005515      IPI       MF</span></span>
<span><span class="co">## 4   IL3RA GO:0005886      NAS       CC</span></span>
<span><span class="co">## 5   IL3RA GO:0005886      TAS       CC</span></span>
<span><span class="co">## 6   IL3RA GO:0009897      IBA       CC</span></span>
<span><span class="co">## 7   IL3RA GO:0019221      IBA       BP</span></span>
<span><span class="co">## 8   IL3RA GO:0019955      IBA       MF</span></span>
<span><span class="co">## 9   IL3RA GO:0036016      IEA       BP</span></span>
<span><span class="co">## 10  IL3RA GO:0038156      IDA       BP</span></span>
<span><span class="co">## 11  IL3RA GO:0043235      IBA       CC</span></span></code></pre>
<p>As shown previously, the GO ID, EVIDENCE code, and ONTOLOGY comprise
the extent of GO information contained in <code>SomaScan.db</code>.
However, we can use the GO ID (in the <code>GO</code> column) to connect
these values to the annotations in <em><a href="https://bioconductor.org/packages/3.18/GO.db" class="external-link">GO.db</a></em>:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">columns</a></span><span class="op">(</span><span class="va">GO.db</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "DEFINITION" "GOID"       "ONTOLOGY"   "TERM"</span></span></code></pre>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">go_defs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/select-SomaDb-method.html">select</a></span><span class="op">(</span><span class="va">GO.db</span>, keys <span class="op">=</span> <span class="va">go_ids</span><span class="op">$</span><span class="va">GO</span>,</span>
<span>                  columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GOID"</span>, <span class="st">"TERM"</span>, <span class="st">"DEFINITION"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'select()' returned many:1 mapping between keys and columns</span></span></code></pre>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">go_defs</span></span></code></pre></div>
<pre><code><span><span class="co">##          GOID                                     TERM</span></span>
<span><span class="co">## 1  GO:0004896               cytokine receptor activity</span></span>
<span><span class="co">## 2  GO:0004912          interleukin-3 receptor activity</span></span>
<span><span class="co">## 3  GO:0005515                          protein binding</span></span>
<span><span class="co">## 4  GO:0005886                          plasma membrane</span></span>
<span><span class="co">## 5  GO:0005886                          plasma membrane</span></span>
<span><span class="co">## 6  GO:0009897         external side of plasma membrane</span></span>
<span><span class="co">## 7  GO:0019221      cytokine-mediated signaling pathway</span></span>
<span><span class="co">## 8  GO:0019955                         cytokine binding</span></span>
<span><span class="co">## 9  GO:0036016       cellular response to interleukin-3</span></span>
<span><span class="co">## 10 GO:0038156 interleukin-3-mediated signaling pathway</span></span>
<span><span class="co">## 11 GO:0043235                         receptor complex</span></span>
<span><span class="co">##                                                                                                                                                                                                  DEFINITION</span></span>
<span><span class="co">## 1                                                                   Combining with a cytokine and transmitting the signal from one side of the membrane to the other to initiate a change in cell activity.</span></span>
<span><span class="co">## 2                                                                Combining with interleukin-3 and transmitting the signal from one side of the membrane to the other to initiate a change in cell activity.</span></span>
<span><span class="co">## 3                                                                                                                                                                                     Binding to a protein.</span></span>
<span><span class="co">## 4                                                     The membrane surrounding a cell that separates the cell from its external environment. It consists of a phospholipid bilayer and associated proteins.</span></span>
<span><span class="co">## 5                                                     The membrane surrounding a cell that separates the cell from its external environment. It consists of a phospholipid bilayer and associated proteins.</span></span>
<span><span class="co">## 6                                                             The leaflet of the plasma membrane that faces away from the cytoplasm and any proteins embedded or anchored in it or attached to its surface.</span></span>
<span><span class="co">## 7       The series of molecular signals initiated by the binding of a cytokine to a receptor on the surface of a cell, and ending with the regulation of a downstream cellular process, e.g. transcription.</span></span>
<span><span class="co">## 8                Binding to a cytokine, any of a group of proteins that function to control the survival, growth and differentiation of tissues and cells, and which have autocrine and paracrine activity.</span></span>
<span><span class="co">## 9                 Any process that results in a change in state or activity of a cell (in terms of movement, secretion, enzyme production, gene expression, etc.) as a result of an interleukin-3 stimulus.</span></span>
<span><span class="co">## 10 The series of molecular signals initiated by interleukin-3 binding to its receptor on the surface of a target cell, and ending with the regulation of a downstream cellular process, e.g. transcription.</span></span>
<span><span class="co">## 11                                                  Any protein complex that undergoes combination with a hormone, neurotransmitter, drug or intracellular messenger to initiate a change in cell function.</span></span></code></pre>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">go_ids</span>, <span class="va">go_defs</span>, by.x <span class="op">=</span> <span class="st">"GO"</span>, by.y <span class="op">=</span> <span class="st">"GOID"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            GO SYMBOL EVIDENCE ONTOLOGY</span></span>
<span><span class="co">## 1  GO:0004896  IL3RA      IBA       MF</span></span>
<span><span class="co">## 2  GO:0004912  IL3RA      IDA       MF</span></span>
<span><span class="co">## 3  GO:0005515  IL3RA      IPI       MF</span></span>
<span><span class="co">## 4  GO:0005886  IL3RA      NAS       CC</span></span>
<span><span class="co">## 5  GO:0005886  IL3RA      NAS       CC</span></span>
<span><span class="co">## 6  GO:0005886  IL3RA      TAS       CC</span></span>
<span><span class="co">## 7  GO:0005886  IL3RA      TAS       CC</span></span>
<span><span class="co">## 8  GO:0009897  IL3RA      IBA       CC</span></span>
<span><span class="co">## 9  GO:0019221  IL3RA      IBA       BP</span></span>
<span><span class="co">## 10 GO:0019955  IL3RA      IBA       MF</span></span>
<span><span class="co">## 11 GO:0036016  IL3RA      IEA       BP</span></span>
<span><span class="co">## 12 GO:0038156  IL3RA      IDA       BP</span></span>
<span><span class="co">## 13 GO:0043235  IL3RA      IBA       CC</span></span>
<span><span class="co">##                                        TERM</span></span>
<span><span class="co">## 1                cytokine receptor activity</span></span>
<span><span class="co">## 2           interleukin-3 receptor activity</span></span>
<span><span class="co">## 3                           protein binding</span></span>
<span><span class="co">## 4                           plasma membrane</span></span>
<span><span class="co">## 5                           plasma membrane</span></span>
<span><span class="co">## 6                           plasma membrane</span></span>
<span><span class="co">## 7                           plasma membrane</span></span>
<span><span class="co">## 8          external side of plasma membrane</span></span>
<span><span class="co">## 9       cytokine-mediated signaling pathway</span></span>
<span><span class="co">## 10                         cytokine binding</span></span>
<span><span class="co">## 11       cellular response to interleukin-3</span></span>
<span><span class="co">## 12 interleukin-3-mediated signaling pathway</span></span>
<span><span class="co">## 13                         receptor complex</span></span>
<span><span class="co">##                                                                                                                                                                                                  DEFINITION</span></span>
<span><span class="co">## 1                                                                   Combining with a cytokine and transmitting the signal from one side of the membrane to the other to initiate a change in cell activity.</span></span>
<span><span class="co">## 2                                                                Combining with interleukin-3 and transmitting the signal from one side of the membrane to the other to initiate a change in cell activity.</span></span>
<span><span class="co">## 3                                                                                                                                                                                     Binding to a protein.</span></span>
<span><span class="co">## 4                                                     The membrane surrounding a cell that separates the cell from its external environment. It consists of a phospholipid bilayer and associated proteins.</span></span>
<span><span class="co">## 5                                                     The membrane surrounding a cell that separates the cell from its external environment. It consists of a phospholipid bilayer and associated proteins.</span></span>
<span><span class="co">## 6                                                     The membrane surrounding a cell that separates the cell from its external environment. It consists of a phospholipid bilayer and associated proteins.</span></span>
<span><span class="co">## 7                                                     The membrane surrounding a cell that separates the cell from its external environment. It consists of a phospholipid bilayer and associated proteins.</span></span>
<span><span class="co">## 8                                                             The leaflet of the plasma membrane that faces away from the cytoplasm and any proteins embedded or anchored in it or attached to its surface.</span></span>
<span><span class="co">## 9       The series of molecular signals initiated by the binding of a cytokine to a receptor on the surface of a cell, and ending with the regulation of a downstream cellular process, e.g. transcription.</span></span>
<span><span class="co">## 10               Binding to a cytokine, any of a group of proteins that function to control the survival, growth and differentiation of tissues and cells, and which have autocrine and paracrine activity.</span></span>
<span><span class="co">## 11                Any process that results in a change in state or activity of a cell (in terms of movement, secretion, enzyme production, gene expression, etc.) as a result of an interleukin-3 stimulus.</span></span>
<span><span class="co">## 12 The series of molecular signals initiated by interleukin-3 binding to its receptor on the surface of a target cell, and ending with the regulation of a downstream cellular process, e.g. transcription.</span></span>
<span><span class="co">## 13                                                  Any protein complex that undergoes combination with a hormone, neurotransmitter, drug or intracellular messenger to initiate a change in cell function.</span></span></code></pre>
<p>Using this workflow, in just two steps we can link annotation
information <em>between</em> annotation package resources (i.e
<code>SomaScan.db</code> &lt;–&gt; <code>GO.db</code>).</p>
<hr>
</div>
</div>
<div class="section level3">
<h3 id="kegg-pathways">KEGG Pathways<a class="anchor" aria-label="anchor" href="#kegg-pathways"></a>
</h3>
<p>Note that the same workflow <em>cannot</em> be performed for KEGG
pathways, due to KEGG’s data sharing policy. Instead, the package <em><a href="https://bioconductor.org/packages/3.18/KEGGREST" class="external-link">KEGGREST</a></em>
must be used. Rather than an annotation database-style package (like
<code>SomaScan.db</code> and <code>GO.db</code>), <em><a href="https://bioconductor.org/packages/3.18/KEGGREST" class="external-link">KEGGREST</a></em>
is a package that provides a client interface in R to the KEGG REST
(REpresentational State Transfer) server. For reference, REST is an
interface that two computer systems can use to securely exchange
information over the internet. Queries made with the <em><a href="https://bioconductor.org/packages/3.18/KEGGREST" class="external-link">KEGGREST</a></em>
package retrieve information directly from the online KEGG database.</p>
<p>Let’s take the same <code>select</code> query as we used for GO, but
modify it to obtain KEGG pathway identifiers instead:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kegg_sel</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/select-SomaDb-method.html">select</a></span><span class="op">(</span><span class="va">SomaScan.db</span>, keys <span class="op">=</span> <span class="st">"CD86"</span>, keytype <span class="op">=</span> <span class="st">"SYMBOL"</span>, </span>
<span>                   columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PROBEID"</span>, <span class="st">"PATH"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'select()' returned 1:many mapping between keys and columns</span></span></code></pre>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kegg_sel</span></span></code></pre></div>
<pre><code><span><span class="co">##    SYMBOL PROBEID  PATH</span></span>
<span><span class="co">## 1    CD86 5337-64 04514</span></span>
<span><span class="co">## 2    CD86 5337-64 04620</span></span>
<span><span class="co">## 3    CD86 5337-64 04672</span></span>
<span><span class="co">## 4    CD86 5337-64 04940</span></span>
<span><span class="co">## 5    CD86 5337-64 05320</span></span>
<span><span class="co">## 6    CD86 5337-64 05322</span></span>
<span><span class="co">## 7    CD86 5337-64 05323</span></span>
<span><span class="co">## 8    CD86 5337-64 05330</span></span>
<span><span class="co">## 9    CD86 5337-64 05332</span></span>
<span><span class="co">## 10   CD86 5337-64 05416</span></span>
<span><span class="co">## 11   CD86 6232-54 04514</span></span>
<span><span class="co">## 12   CD86 6232-54 04620</span></span>
<span><span class="co">## 13   CD86 6232-54 04672</span></span>
<span><span class="co">## 14   CD86 6232-54 04940</span></span>
<span><span class="co">## 15   CD86 6232-54 05320</span></span>
<span><span class="co">## 16   CD86 6232-54 05322</span></span>
<span><span class="co">## 17   CD86 6232-54 05323</span></span>
<span><span class="co">## 18   CD86 6232-54 05330</span></span>
<span><span class="co">## 19   CD86 6232-54 05332</span></span>
<span><span class="co">## 20   CD86 6232-54 05416</span></span></code></pre>
<p>We can use the identifiers in the “PATH” column to query the KEGG
database using <code><a href="https://rdrr.io/pkg/KEGGREST/man/keggGet.html" class="external-link">KEGGREST::keggGet()</a></code>:</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Add prefix indicating species (hsa = Homo sapiens)</span></span>
<span><span class="va">hsa_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"hsa"</span>, <span class="va">kegg_sel</span><span class="op">$</span><span class="va">PATH</span><span class="op">)</span></span>
<span></span>
<span><span class="va">kegg_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/KEGGREST/man/keggGet.html" class="external-link">keggGet</a></span><span class="op">(</span>dbentries <span class="op">=</span> <span class="va">hsa_names</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="va">hsa_names</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10L</span><span class="op">]</span><span class="op">)</span> <span class="co"># Setting names for results list</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning in keggGet(dbentries = hsa_names): More than 10 inputs supplied,</span></span>
<span><span class="co">## only the first 10 results will be returned.</span></span></code></pre>
<p>Because so much information is returned by <code><a href="https://rdrr.io/pkg/KEGGREST/man/keggGet.html" class="external-link">keggGet()</a></code>, a
maximum number of 10 entries are allowed. Input exceeding 10 entries
will be truncated, and only the first 10 results will be returned (as
indicated in the warning message above). Let’s take a look at what was
returned for each KEGG pathway:</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">kegg_res</span><span class="op">$</span><span class="va">hsa04514</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## List of 12</span></span>
<span><span class="co">##  $ ENTRY      : Named chr "hsa04514"</span></span>
<span><span class="co">##   ..- attr(*, "names")= chr "Pathway"</span></span>
<span><span class="co">##  $ NAME       : chr "Cell adhesion molecules - Homo sapiens (human)"</span></span>
<span><span class="co">##  $ DESCRIPTION: chr "Cell adhesion molecules (CAMs) are (glyco)proteins expressed on the cell surface and play a critical role in a "| __truncated__</span></span>
<span><span class="co">##  $ CLASS      : chr "Environmental Information Processing; Signaling molecules and interaction"</span></span>
<span><span class="co">##  $ PATHWAY_MAP: Named chr "Cell adhesion molecules"</span></span>
<span><span class="co">##   ..- attr(*, "names")= chr "hsa04514"</span></span>
<span><span class="co">##  $ DRUG       : chr [1:210] "D02800" "Alefacept (USAN/INN)" "D02811" "Alicaforsen sodium (USAN)" ...</span></span>
<span><span class="co">##  $ DBLINKS    : chr "GO: 0050839"</span></span>
<span><span class="co">##  $ ORGANISM   : Named chr "NA  Homo sapiens (human) [GN:hsa]"</span></span>
<span><span class="co">##   ..- attr(*, "names")= chr "Homo sapiens (human) [GN:hsa]"</span></span>
<span><span class="co">##  $ GENE       : chr [1:316] "965" "CD58; CD58 molecule [KO:K06492]" "914" "CD2; CD2 molecule [KO:K06449]" ...</span></span>
<span><span class="co">##  $ REL_PATHWAY: Named chr [1:5] "Adherens junction" "Tight junction" "Complement and coagulation cascades" "T cell receptor signaling pathway" ...</span></span>
<span><span class="co">##   ..- attr(*, "names")= chr [1:5] "hsa04520" "hsa04530" "hsa04610" "hsa04660" ...</span></span>
<span><span class="co">##  $ KO_PATHWAY : chr "ko04514"</span></span>
<span><span class="co">##  $ REFERENCE  :List of 25</span></span>
<span><span class="co">##   ..$ :List of 4</span></span>
<span><span class="co">##   .. ..$ REFERENCE: chr "PMID:14690046"</span></span>
<span><span class="co">##   .. ..$ AUTHORS  : chr "Barclay AN."</span></span>
<span><span class="co">##   .. ..$ TITLE    : chr "Membrane proteins with immunoglobulin-like domains--a master superfamily of interaction molecules."</span></span>
<span><span class="co">##   .. ..$ JOURNAL  : chr [1:2] "Semin Immunol 15:215-23 (2003)" "DOI:10.1016/S1044-5323(03)00047-2"</span></span>
<span><span class="co">##   ..$ :List of 4</span></span>
<span><span class="co">##   .. ..$ REFERENCE: chr "PMID:11910893"</span></span>
<span><span class="co">##   .. ..$ AUTHORS  : chr "Sharpe AH, Freeman GJ."</span></span>
<span><span class="co">##   .. ..$ TITLE    : chr "The B7-CD28 superfamily."</span></span>
<span><span class="co">##   .. ..$ JOURNAL  : chr [1:2] "Nat Rev Immunol 2:116-26 (2002)" "DOI:10.1038/nri727"</span></span>
<span><span class="co">##   ..$ :List of 4</span></span>
<span><span class="co">##   .. ..$ REFERENCE: chr "PMID:9597126"</span></span>
<span><span class="co">##   .. ..$ AUTHORS  : chr "Grewal IS, Flavell RA."</span></span>
<span><span class="co">##   .. ..$ TITLE    : chr "CD40 and CD154 in cell-mediated immunity."</span></span>
<span><span class="co">##   .. ..$ JOURNAL  : chr [1:2] "Annu Rev Immunol 16:111-35 (1998)" "DOI:10.1146/annurev.immunol.16.1.111"</span></span>
<span><span class="co">##   ..$ :List of 4</span></span>
<span><span class="co">##   .. ..$ REFERENCE: chr "PMID:16034094"</span></span>
<span><span class="co">##   .. ..$ AUTHORS  : chr "Dardalhon V, Schubart AS, Reddy J, Meyers JH, Monney L, Sabatos CA, Ahuja R, Nguyen K, Freeman GJ, Greenfield E"| __truncated__</span></span>
<span><span class="co">##   .. ..$ TITLE    : chr "CD226 is specifically expressed on the surface of Th1 cells and regulates their expansion and effector functions."</span></span>
<span><span class="co">##   .. ..$ JOURNAL  : chr [1:2] "J Immunol 175:1558-65 (2005)" "DOI:10.4049/jimmunol.175.3.1558"</span></span>
<span><span class="co">##   ..$ :List of 4</span></span>
<span><span class="co">##   .. ..$ REFERENCE: chr "PMID:12234363"</span></span>
<span><span class="co">##   .. ..$ AUTHORS  : chr "Montoya MC, Sancho D, Vicente-Manzanares M, Sanchez-Madrid F."</span></span>
<span><span class="co">##   .. ..$ TITLE    : chr "Cell adhesion and polarity during immune interactions."</span></span>
<span><span class="co">##   .. ..$ JOURNAL  : chr [1:2] "Immunol Rev 186:68-82 (2002)" "DOI:10.1034/j.1600-065X.2002.18607.x"</span></span>
<span><span class="co">##   ..$ :List of 4</span></span>
<span><span class="co">##   .. ..$ REFERENCE: chr "PMID:15071551"</span></span>
<span><span class="co">##   .. ..$ AUTHORS  : chr "Dejana E."</span></span>
<span><span class="co">##   .. ..$ TITLE    : chr "Endothelial cell-cell junctions: happy together."</span></span>
<span><span class="co">##   .. ..$ JOURNAL  : chr [1:2] "Nat Rev Mol Cell Biol 5:261-70 (2004)" "DOI:10.1038/nrm1357"</span></span>
<span><span class="co">##   ..$ :List of 4</span></span>
<span><span class="co">##   .. ..$ REFERENCE: chr "PMID:14519386"</span></span>
<span><span class="co">##   .. ..$ AUTHORS  : chr "Bazzoni G."</span></span>
<span><span class="co">##   .. ..$ TITLE    : chr "The JAM family of junctional adhesion molecules."</span></span>
<span><span class="co">##   .. ..$ JOURNAL  : chr [1:2] "Curr Opin Cell Biol 15:525-30 (2003)" "DOI:10.1016/S0955-0674(03)00104-2"</span></span>
<span><span class="co">##   ..$ :List of 4</span></span>
<span><span class="co">##   .. ..$ REFERENCE: chr "PMID:10798271"</span></span>
<span><span class="co">##   .. ..$ AUTHORS  : chr "Becker BF, Heindl B, Kupatt C, Zahler S."</span></span>
<span><span class="co">##   .. ..$ TITLE    : chr "Endothelial function and hemostasis."</span></span>
<span><span class="co">##   .. ..$ JOURNAL  : chr [1:2] "Z Kardiol 89:160-7 (2000)" "DOI:10.1007/PL00007320"</span></span>
<span><span class="co">##   ..$ :List of 4</span></span>
<span><span class="co">##   .. ..$ REFERENCE: chr "PMID:9150551"</span></span>
<span><span class="co">##   .. ..$ AUTHORS  : chr "Elangbam CS, Qualls CW Jr, Dahlgren RR."</span></span>
<span><span class="co">##   .. ..$ TITLE    : chr "Cell adhesion molecules--update."</span></span>
<span><span class="co">##   .. ..$ JOURNAL  : chr [1:2] "Vet Pathol 34:61-73 (1997)" "DOI:10.1177/030098589703400113"</span></span>
<span><span class="co">##   ..$ :List of 4</span></span>
<span><span class="co">##   .. ..$ REFERENCE: chr "PMID:12810109"</span></span>
<span><span class="co">##   .. ..$ AUTHORS  : chr "Muller WA."</span></span>
<span><span class="co">##   .. ..$ TITLE    : chr "Leukocyte-endothelial-cell interactions in leukocyte transmigration and the inflammatory response."</span></span>
<span><span class="co">##   .. ..$ JOURNAL  : chr [1:2] "Trends Immunol 24:327-34 (2003)" "DOI:10.1016/S1471-4906(03)00117-0"</span></span>
<span><span class="co">##   ..$ :List of 4</span></span>
<span><span class="co">##   .. ..$ REFERENCE: chr "PMID:14519398"</span></span>
<span><span class="co">##   .. ..$ AUTHORS  : chr "Yamagata M, Sanes JR, Weiner JA."</span></span>
<span><span class="co">##   .. ..$ TITLE    : chr "Synaptic adhesion molecules."</span></span>
<span><span class="co">##   .. ..$ JOURNAL  : chr [1:2] "Curr Opin Cell Biol 15:621-32 (2003)" "DOI:10.1016/S0955-0674(03)00107-8"</span></span>
<span><span class="co">##   ..$ :List of 4</span></span>
<span><span class="co">##   .. ..$ REFERENCE: chr "PMID:15882774"</span></span>
<span><span class="co">##   .. ..$ AUTHORS  : chr "Ethell IM, Pasquale EB."</span></span>
<span><span class="co">##   .. ..$ TITLE    : chr "Molecular mechanisms of dendritic spine development and remodeling."</span></span>
<span><span class="co">##   .. ..$ JOURNAL  : chr [1:2] "Prog Neurobiol 75:161-205 (2005)" "DOI:10.1016/j.pneurobio.2005.02.003"</span></span>
<span><span class="co">##   ..$ :List of 4</span></span>
<span><span class="co">##   .. ..$ REFERENCE: chr "PMID:11050419"</span></span>
<span><span class="co">##   .. ..$ AUTHORS  : chr "Benson DL, Schnapp LM, Shapiro L, Huntley GW."</span></span>
<span><span class="co">##   .. ..$ TITLE    : chr "Making memories stick: cell-adhesion molecules in synaptic plasticity."</span></span>
<span><span class="co">##   .. ..$ JOURNAL  : chr [1:2] "Trends Cell Biol 10:473-82 (2000)" "DOI:10.1016/S0962-8924(00)01838-9"</span></span>
<span><span class="co">##   ..$ :List of 4</span></span>
<span><span class="co">##   .. ..$ REFERENCE: chr "PMID:11860281"</span></span>
<span><span class="co">##   .. ..$ AUTHORS  : chr "Rosdahl JA, Mourton TL, Brady-Kalnay SM."</span></span>
<span><span class="co">##   .. ..$ TITLE    : chr "Protein kinase C delta (PKCdelta) is required for protein tyrosine phosphatase mu (PTPmu)-dependent neurite outgrowth."</span></span>
<span><span class="co">##   .. ..$ JOURNAL  : chr [1:2] "Mol Cell Neurosci 19:292-306 (2002)" "DOI:10.1006/mcne.2001.1071"</span></span>
<span><span class="co">##   ..$ :List of 4</span></span>
<span><span class="co">##   .. ..$ REFERENCE: chr "PMID:10964748"</span></span>
<span><span class="co">##   .. ..$ AUTHORS  : chr "Dunican DJ, Doherty P."</span></span>
<span><span class="co">##   .. ..$ TITLE    : chr "The generation of localized calcium rises mediated by cell adhesion molecules and their role in neuronal growth cone motility."</span></span>
<span><span class="co">##   .. ..$ JOURNAL  : chr [1:2] "Mol Cell Biol Res Commun 3:255-63 (2000)" "DOI:10.1006/mcbr.2000.0225"</span></span>
<span><span class="co">##   ..$ :List of 4</span></span>
<span><span class="co">##   .. ..$ REFERENCE: chr "PMID:12367625"</span></span>
<span><span class="co">##   .. ..$ AUTHORS  : chr "Girault JA, Peles E."</span></span>
<span><span class="co">##   .. ..$ TITLE    : chr "Development of nodes of Ranvier."</span></span>
<span><span class="co">##   .. ..$ JOURNAL  : chr [1:2] "Curr Opin Neurobiol 12:476-85 (2002)" "DOI:10.1016/S0959-4388(02)00370-7"</span></span>
<span><span class="co">##   ..$ :List of 4</span></span>
<span><span class="co">##   .. ..$ REFERENCE: chr "PMID:10664064"</span></span>
<span><span class="co">##   .. ..$ AUTHORS  : chr "Arroyo EJ, Scherer SS."</span></span>
<span><span class="co">##   .. ..$ TITLE    : chr "On the molecular architecture of myelinated fibers."</span></span>
<span><span class="co">##   .. ..$ JOURNAL  : chr [1:2] "Histochem Cell Biol 113:1-18 (2000)" "DOI:10.1007/s004180050001"</span></span>
<span><span class="co">##   ..$ :List of 4</span></span>
<span><span class="co">##   .. ..$ REFERENCE: chr "PMID:14556710"</span></span>
<span><span class="co">##   .. ..$ AUTHORS  : chr "Salzer JL."</span></span>
<span><span class="co">##   .. ..$ TITLE    : chr "Polarized domains of myelinated axons."</span></span>
<span><span class="co">##   .. ..$ JOURNAL  : chr [1:2] "Neuron 40:297-318 (2003)" "DOI:10.1016/S0896-6273(03)00628-7"</span></span>
<span><span class="co">##   ..$ :List of 4</span></span>
<span><span class="co">##   .. ..$ REFERENCE: chr "PMID:15561584"</span></span>
<span><span class="co">##   .. ..$ AUTHORS  : chr "Irie K, Shimizu K, Sakisaka T, Ikeda W, Takai Y."</span></span>
<span><span class="co">##   .. ..$ TITLE    : chr "Roles and modes of action of nectins in cell-cell adhesion."</span></span>
<span><span class="co">##   .. ..$ JOURNAL  : chr [1:2] "Semin Cell Dev Biol 15:643-56 (2004)" "DOI:10.1016/j.semcdb.2004.09.002"</span></span>
<span><span class="co">##   ..$ :List of 4</span></span>
<span><span class="co">##   .. ..$ REFERENCE: chr "PMID:15551862"</span></span>
<span><span class="co">##   .. ..$ AUTHORS  : chr "Nakanishi H, Takai Y."</span></span>
<span><span class="co">##   .. ..$ TITLE    : chr "Roles of nectins in cell adhesion, migration and polarization."</span></span>
<span><span class="co">##   .. ..$ JOURNAL  : chr [1:2] "Biol Chem 385:885-92 (2004)" "DOI:10.1515/BC.2004.116"</span></span>
<span><span class="co">##   ..$ :List of 4</span></span>
<span><span class="co">##   .. ..$ REFERENCE: chr "PMID:15115723"</span></span>
<span><span class="co">##   .. ..$ AUTHORS  : chr "Siu MK, Cheng CY."</span></span>
<span><span class="co">##   .. ..$ TITLE    : chr "Extracellular matrix: recent advances on its role in junction dynamics in the seminiferous epithelium during spermatogenesis."</span></span>
<span><span class="co">##   .. ..$ JOURNAL  : chr [1:2] "Biol Reprod 71:375-91 (2004)" "DOI:10.1095/biolreprod.104.028225"</span></span>
<span><span class="co">##   ..$ :List of 4</span></span>
<span><span class="co">##   .. ..$ REFERENCE: chr "PMID:15056568"</span></span>
<span><span class="co">##   .. ..$ AUTHORS  : chr "Lee NP, Cheng CY."</span></span>
<span><span class="co">##   .. ..$ TITLE    : chr "Adaptors, junction dynamics, and spermatogenesis."</span></span>
<span><span class="co">##   .. ..$ JOURNAL  : chr [1:2] "Biol Reprod 71:392-404 (2004)" "DOI:10.1095/biolreprod.104.027268"</span></span>
<span><span class="co">##   ..$ :List of 4</span></span>
<span><span class="co">##   .. ..$ REFERENCE: chr "PMID:15728677"</span></span>
<span><span class="co">##   .. ..$ AUTHORS  : chr "Inagaki M, Irie K, Ishizaki H, Tanaka-Okamoto M, Morimoto K, Inoue E, Ohtsuka T, Miyoshi J, Takai Y."</span></span>
<span><span class="co">##   .. ..$ TITLE    : chr "Roles of cell-adhesion molecules nectin 1 and nectin 3 in ciliary body development."</span></span>
<span><span class="co">##   .. ..$ JOURNAL  : chr [1:2] "Development 132:1525-37 (2005)" "DOI:10.1242/dev.01697"</span></span>
<span><span class="co">##   ..$ :List of 4</span></span>
<span><span class="co">##   .. ..$ REFERENCE: chr "PMID:12500939"</span></span>
<span><span class="co">##   .. ..$ AUTHORS  : chr "Marthiens V, Gavard J, Lambert M, Mege RM."</span></span>
<span><span class="co">##   .. ..$ TITLE    : chr "Cadherin-based cell adhesion in neuromuscular development."</span></span>
<span><span class="co">##   .. ..$ JOURNAL  : chr [1:2] "Biol Cell 94:315-26 (2002)" "DOI:10.1016/S0248-4900(02)00005-9"</span></span>
<span><span class="co">##   ..$ :List of 4</span></span>
<span><span class="co">##   .. ..$ REFERENCE: chr "PMID:15923648"</span></span>
<span><span class="co">##   .. ..$ AUTHORS  : chr "Krauss RS, Cole F, Gaio U, Takaesu G, Zhang W, Kang JS."</span></span>
<span><span class="co">##   .. ..$ TITLE    : chr "Close encounters: regulation of vertebrate skeletal myogenesis by cell-cell contact."</span></span>
<span><span class="co">##   .. ..$ JOURNAL  : chr [1:2] "J Cell Sci 118:2355-62 (2005)" "DOI:10.1242/jcs.02397"</span></span></code></pre>
<p>Some additional data manipulation will be required to extract the
desired information from the results of <code><a href="https://rdrr.io/pkg/KEGGREST/man/keggGet.html" class="external-link">keggGet()</a></code>. Let’s
just extract the pathway name (<code>NAME</code>):</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kegg_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">vapply</a></span><span class="op">(</span><span class="va">kegg_res</span>, <span class="va">`[[`</span>, i <span class="op">=</span> <span class="st">"NAME"</span>, <span class="st">""</span>, USE.NAMES <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">kegg_names</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "Cell adhesion molecules - Homo sapiens (human)"                     </span></span>
<span><span class="co">##  [2] "Toll-like receptor signaling pathway - Homo sapiens (human)"        </span></span>
<span><span class="co">##  [3] "Intestinal immune network for IgA production - Homo sapiens (human)"</span></span>
<span><span class="co">##  [4] "Type I diabetes mellitus - Homo sapiens (human)"                    </span></span>
<span><span class="co">##  [5] "Autoimmune thyroid disease - Homo sapiens (human)"                  </span></span>
<span><span class="co">##  [6] "Systemic lupus erythematosus - Homo sapiens (human)"                </span></span>
<span><span class="co">##  [7] "Rheumatoid arthritis - Homo sapiens (human)"                        </span></span>
<span><span class="co">##  [8] "Allograft rejection - Homo sapiens (human)"                         </span></span>
<span><span class="co">##  [9] "Graft-versus-host disease - Homo sapiens (human)"                   </span></span>
<span><span class="co">## [10] "Viral myocarditis - Homo sapiens (human)"</span></span></code></pre>
<p>Now we can append this vector to our original results from
<code>select</code>:</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kegg_sel</span><span class="op">$</span><span class="va">PATHNAME</span> <span class="op">&lt;-</span> <span class="va">kegg_names</span></span>
<span></span>
<span><span class="va">kegg_sel</span></span></code></pre></div>
<pre><code><span><span class="co">##    SYMBOL PROBEID  PATH</span></span>
<span><span class="co">## 1    CD86 5337-64 04514</span></span>
<span><span class="co">## 2    CD86 5337-64 04620</span></span>
<span><span class="co">## 3    CD86 5337-64 04672</span></span>
<span><span class="co">## 4    CD86 5337-64 04940</span></span>
<span><span class="co">## 5    CD86 5337-64 05320</span></span>
<span><span class="co">## 6    CD86 5337-64 05322</span></span>
<span><span class="co">## 7    CD86 5337-64 05323</span></span>
<span><span class="co">## 8    CD86 5337-64 05330</span></span>
<span><span class="co">## 9    CD86 5337-64 05332</span></span>
<span><span class="co">## 10   CD86 5337-64 05416</span></span>
<span><span class="co">## 11   CD86 6232-54 04514</span></span>
<span><span class="co">## 12   CD86 6232-54 04620</span></span>
<span><span class="co">## 13   CD86 6232-54 04672</span></span>
<span><span class="co">## 14   CD86 6232-54 04940</span></span>
<span><span class="co">## 15   CD86 6232-54 05320</span></span>
<span><span class="co">## 16   CD86 6232-54 05322</span></span>
<span><span class="co">## 17   CD86 6232-54 05323</span></span>
<span><span class="co">## 18   CD86 6232-54 05330</span></span>
<span><span class="co">## 19   CD86 6232-54 05332</span></span>
<span><span class="co">## 20   CD86 6232-54 05416</span></span>
<span><span class="co">##                                                               PATHNAME</span></span>
<span><span class="co">## 1                       Cell adhesion molecules - Homo sapiens (human)</span></span>
<span><span class="co">## 2          Toll-like receptor signaling pathway - Homo sapiens (human)</span></span>
<span><span class="co">## 3  Intestinal immune network for IgA production - Homo sapiens (human)</span></span>
<span><span class="co">## 4                      Type I diabetes mellitus - Homo sapiens (human)</span></span>
<span><span class="co">## 5                    Autoimmune thyroid disease - Homo sapiens (human)</span></span>
<span><span class="co">## 6                  Systemic lupus erythematosus - Homo sapiens (human)</span></span>
<span><span class="co">## 7                          Rheumatoid arthritis - Homo sapiens (human)</span></span>
<span><span class="co">## 8                           Allograft rejection - Homo sapiens (human)</span></span>
<span><span class="co">## 9                     Graft-versus-host disease - Homo sapiens (human)</span></span>
<span><span class="co">## 10                            Viral myocarditis - Homo sapiens (human)</span></span>
<span><span class="co">## 11                      Cell adhesion molecules - Homo sapiens (human)</span></span>
<span><span class="co">## 12         Toll-like receptor signaling pathway - Homo sapiens (human)</span></span>
<span><span class="co">## 13 Intestinal immune network for IgA production - Homo sapiens (human)</span></span>
<span><span class="co">## 14                     Type I diabetes mellitus - Homo sapiens (human)</span></span>
<span><span class="co">## 15                   Autoimmune thyroid disease - Homo sapiens (human)</span></span>
<span><span class="co">## 16                 Systemic lupus erythematosus - Homo sapiens (human)</span></span>
<span><span class="co">## 17                         Rheumatoid arthritis - Homo sapiens (human)</span></span>
<span><span class="co">## 18                          Allograft rejection - Homo sapiens (human)</span></span>
<span><span class="co">## 19                    Graft-versus-host disease - Homo sapiens (human)</span></span>
<span><span class="co">## 20                            Viral myocarditis - Homo sapiens (human)</span></span></code></pre>
<p>Other pieces of information can be extracted to the list and reduced
to a character vector or used to build a data frame, which can then be
appended to or merged similar to the pathway name in the code chunks
above. For more details about what can be done with the package, see
<em><a href="https://bioconductor.org/packages/3.18/KEGGREST" class="external-link">KEGGREST</a></em>.</p>
</div>
</div>
<div class="section level2">
<h2 id="positional-annotation">Positional Annotation<a class="anchor" aria-label="anchor" href="#positional-annotation"></a>
</h2>
<p>Similar to the extended GO annotation in the previous section,
positional annotation cannot currently be performed within
<code>SomaScan.db</code>. <code>SomaScan.db</code> is a platform-centric
annotation package, built around the probes of the SomaScan protein
assay, and positional annotation is not within its scope. However, it
<em>is</em> possible to retrieve positional annotations by linking to
other Bioconductor annotation resources, which can then be combined with
<code>SomaScan.db</code> in a two-step process (similar to above). The
first step uses <code>SomaScan.db</code> to retrieve gene-level
information corresponding to SomaScan analytes; the second requires a
human transcriptome or organism-centric annotation package to retrieve
the desired chromosomal locations.</p>
<p>We will provide a brief example of this using the popular
organism-centric package, <em><a href="https://bioconductor.org/packages/3.18/EnsDb.Hsapiens.v75" class="external-link">EnsDb.Hsapiens.v75</a></em>,
which contains a database of human annotations derived from
<code>Ensembl release 75</code>. However, this procedure can also be
performed using transcriptome-centric annotation packages like <em><a href="https://bioconductor.org/packages/3.18/TxDb.Hsapiens.UCSC.hg19.knownGene" class="external-link">TxDb.Hsapiens.UCSC.hg19.knownGene</a></em>.</p>
<p>Let’s say we are interested in collecting position information
associated with the protein target corresponding to
<code>SeqId = 11138-16</code>. First, we must determine which gene this
<code>SeqId</code> maps to:</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pos_sel</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/select-SomaDb-method.html">select</a></span><span class="op">(</span><span class="va">SomaScan.db</span>, <span class="st">"11138-16"</span>, columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SYMBOL"</span>, <span class="st">"GENENAME"</span>, </span>
<span>                                                       <span class="st">"ENTREZID"</span>, <span class="st">"ENSEMBL"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'select()' returned 1:1 mapping between keys and columns</span></span></code></pre>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pos_sel</span></span></code></pre></div>
<pre><code><span><span class="co">##    PROBEID SYMBOL                           GENENAME ENTREZID</span></span>
<span><span class="co">## 1 11138-16  RUNX3 RUNX family transcription factor 3      864</span></span>
<span><span class="co">##           ENSEMBL</span></span>
<span><span class="co">## 1 ENSG00000020633</span></span></code></pre>
<p>We now know this probe targets protein encoded by the RUNX3 gene. We
can use <em><a href="https://bioconductor.org/packages/3.18/EnsDb.Hsapiens.v75" class="external-link">EnsDb.Hsapiens.v75</a></em>
to retrieve positional information about RUNX3, like which chromosome
the RUNX3 is on, its start and stop position, and how many exons it has
(at the time of Ensembl’s <code>v75</code> release):</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Install package from Bioconductor, if not already installed</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st">"EnsDb.Hsapiens.v75"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"EnsDb.Hsapiens.v75"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># The central keys of the organism-level database are the Ensembl gene ID</span></span>
<span><span class="fu"><a href="../reference/keys-SomaDb-method.html">keys</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v75</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10L</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Also contains the Ensembl gene ID, so this column can be used for merging</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"ENSEMBL"</span>, <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">columns</a></span><span class="op">(</span><span class="va">SomaScan.db</span><span class="op">)</span>, value <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># These columns will inform us as to what positional information we can </span></span>
<span><span class="co"># retrieve from the organism-level database</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">columns</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v75</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Build a query to retrieve the prot IDs and start/stop pos of protein domains</span></span>
<span><span class="va">pos_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/select-SomaDb-method.html">select</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v75</span>, keys <span class="op">=</span> <span class="st">"ENSG00000020633"</span>, </span>
<span>                  columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GENEBIOTYPE"</span>, <span class="st">"SEQCOORDSYSTEM"</span>, <span class="st">"GENEID"</span>, </span>
<span>                              <span class="st">"PROTEINID"</span>, <span class="st">"PROTDOMSTART"</span>, <span class="st">"PROTDOMEND"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Merge back into `pos_sel` using the "GENEID" column</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">pos_sel</span>, <span class="va">pos_res</span>, by.x <span class="op">=</span> <span class="st">"ENSEMBL"</span>, by.y <span class="op">=</span> <span class="st">"GENEID"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="functional-group-representation">Functional Group Representation<a class="anchor" aria-label="anchor" href="#functional-group-representation"></a>
</h2>
<p>As mentioned in the Introductory Vignette
(<code><a href="../articles/SomaScan-db.html">vignette("SomaScan-db", package = "SomaScan.db")</a></code>), the
<code>SomaScan.db</code> annotation database can be queried using values
other than the central database key, the <code>SeqId</code> (i.e. the
“PROBEID” column). This section will describe additional methods of
retrieving information from the database without using the
<code>SeqId</code>.</p>
<hr>
<div class="section level3">
<h3 id="go-term-coverage">GO Term Coverage<a class="anchor" aria-label="anchor" href="#go-term-coverage"></a>
</h3>
<p>The annotations in <code>SomaScan.db</code> can be used to answer
general questions about SomaScan, without the need for a SomaScan
dataset/ADAT file as a starting point. For example, if one were
interested in proteins involved in cancer progression and metastasis
(and therefore cell adhesion), is the SomaScan menu capable of measuring
proteins involved in cell adhesion? If so, how many of these proteins
can be measured with SomaScan?</p>
<p>We can answer this by examining the coverage of the GO term “cell
adhesion” in both the 5k and 7k SomaScan menus. We don’t need the GO
identifier to get started, as that information can be retrieved from
<code>GO.db</code> using the name of the term as the key:</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/select-SomaDb-method.html">select</a></span><span class="op">(</span><span class="va">GO.db</span>, keys <span class="op">=</span> <span class="st">"cell adhesion"</span>, keytype <span class="op">=</span> <span class="st">"TERM"</span>, </span>
<span>       columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GOID"</span>, <span class="st">"TERM"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'select()' returned 1:1 mapping between keys and columns</span></span></code></pre>
<pre><code><span><span class="co">##            TERM       GOID</span></span>
<span><span class="co">## 1 cell adhesion GO:0007155</span></span></code></pre>
<p>Now that we have the GO ID, we can search in <code>SomaScan.db</code>
to determine how many <code>SeqIds</code> are associated with cell
adhesion.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cellAd_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/select-SomaDb-method.html">select</a></span><span class="op">(</span><span class="va">SomaScan.db</span>, keys <span class="op">=</span> <span class="st">"GO:0007155"</span>, keytype <span class="op">=</span> <span class="st">"GO"</span>,</span>
<span>                     columns <span class="op">=</span> <span class="st">"PROBEID"</span>, <span class="st">"UNIPROTID"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'select()' returned 1:many mapping between keys and columns</span></span></code></pre>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">cellAd_ids</span>, n <span class="op">=</span> <span class="fl">10L</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            GO   PROBEID EVIDENCE ONTOLOGY</span></span>
<span><span class="co">## 1  GO:0007155  10037-98      IBA       BP</span></span>
<span><span class="co">## 2  GO:0007155  10511-10      IEA       BP</span></span>
<span><span class="co">## 3  GO:0007155  10521-10      IEA       BP</span></span>
<span><span class="co">## 4  GO:0007155  10539-30      IEA       BP</span></span>
<span><span class="co">## 5  GO:0007155  10558-26      IBA       BP</span></span>
<span><span class="co">## 6  GO:0007155   10702-1      IBA       BP</span></span>
<span><span class="co">## 7  GO:0007155 10748-216      IBA       BP</span></span>
<span><span class="co">## 8  GO:0007155 10907-116      IEA       BP</span></span>
<span><span class="co">## 9  GO:0007155  10980-11      IDA       BP</span></span>
<span><span class="co">## 10 GO:0007155  11067-13      NAS       BP</span></span></code></pre>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Total number of SeqIds associated with cell adhesion</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">cellAd_ids</span><span class="op">$</span><span class="va">PROBEID</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 377</span></span></code></pre>
<p>There are 377 unique <code>SeqIds</code> associated with the “cell
adhesion” GO term (<em>unique</em> is important here because the data
frame above may contain multiple entries per <code>SeqId</code>, due to
the “EVIDENCE” column). There are 7267 total <code>SeqIds</code> in the
<code>SomaScan.db</code> database, so 5.19% of keys in the database are
associated with cell adhesion.</p>
<p>How many of the <em>total</em> proteins in the cell adhesion GO term
are covered by the SomaScan menu? To answer this question, we first must
use another annotation package, <code>org.Hs.eg.db</code>, to retrieve a
list of all human UniProt IDs associated with the “cell adhesion” GO
term.</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cellAd_prots</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/select-SomaDb-method.html">select</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span>, </span>
<span>                       keys <span class="op">=</span> <span class="st">"GO:0007155"</span>, </span>
<span>                       keytype <span class="op">=</span> <span class="st">"GO"</span>, </span>
<span>                       columns <span class="op">=</span> <span class="st">"UNIPROT"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'select()' returned 1:many mapping between keys and columns</span></span></code></pre>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Again, we take the unique set of proteins</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">cellAd_prots</span><span class="op">$</span><span class="va">UNIPROT</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 875</span></span></code></pre>
<p>The GO term <code>GO:0007155</code> (cell adhesion) contains a total
of 875 unique human UniProt IDs. Now we can check to see how many of
these are covered by the SomaScan menu by searching for the proteins in
<code>SomaScan.db</code> with <code>select</code>:</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cellAd_covProts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/select-SomaDb-method.html">select</a></span><span class="op">(</span><span class="va">SomaScan.db</span>, keys <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">cellAd_prots</span><span class="op">$</span><span class="va">UNIPROT</span><span class="op">)</span>,</span>
<span>                          keytype <span class="op">=</span> <span class="st">"UNIPROT"</span>, columns <span class="op">=</span> <span class="st">"PROBEID"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'select()' returned 1:many mapping between keys and columns</span></span></code></pre>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">cellAd_covProts</span>, n <span class="op">=</span> <span class="fl">20L</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       UNIPROT  PROBEID</span></span>
<span><span class="co">## 1      P42684  3342-76</span></span>
<span><span class="co">## 2      P42684  5261-13</span></span>
<span><span class="co">## 3      P22303 10980-11</span></span>
<span><span class="co">## 4      P22303 15553-22</span></span>
<span><span class="co">## 5      P35609 9844-138</span></span>
<span><span class="co">## 6  A0A0S2Z381 19751-21</span></span>
<span><span class="co">## 7      P00813 19751-21</span></span>
<span><span class="co">## 8      F5GWI4 19751-21</span></span>
<span><span class="co">## 9      E7ENV9  3280-49</span></span>
<span><span class="co">## 10     E7EX88  3280-49</span></span>
<span><span class="co">## 11 A0A1U9X785  4125-52</span></span>
<span><span class="co">## 12     B4DNX3  4125-52</span></span>
<span><span class="co">## 13     Q15109  4125-52</span></span>
<span><span class="co">## 14     Q13740   5451-1</span></span>
<span><span class="co">## 15     B3KNN9   5451-1</span></span>
<span><span class="co">## 16     Q546D7     &lt;NA&gt;</span></span>
<span><span class="co">## 17     Q9NP70     &lt;NA&gt;</span></span>
<span><span class="co">## 18     P02760  15453-3</span></span>
<span><span class="co">## 19     Q99217     &lt;NA&gt;</span></span>
<span><span class="co">## 20     Q02410     &lt;NA&gt;</span></span></code></pre>
<p><code>select</code> will return an <code>NA</code> value if a key is
not found in the database. As seen above, some proteins in
<code>GO:0007155</code> do not map to a <code>SeqId</code> in
<code>SomaScan.db</code>. To get an accurate count of the proteins that
<em>do</em> map to a <code>SeqId</code>, we must remove the unmapped
proteins by filtering out rows with <code>NA</code> values:</p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cellAd_covProts</span> <span class="op">&lt;-</span> <span class="va">cellAd_covProts</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">cellAd_covProts</span><span class="op">$</span><span class="va">PROBEID</span><span class="op">)</span>,<span class="op">]</span></span>
<span></span>
<span><span class="va">cellAd_covIDs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">cellAd_covProts</span><span class="op">$</span><span class="va">UNIPROT</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">cellAd_covIDs</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 542</span></span></code></pre>
<p>We removed duplicates from the list of proteins provided as keys, to
get a final count of 542 proteins (61.94%) from the “cell adhesion” GO
term that are covered by the SomaScan menu.</p>
<p>Does this number differ between versions of the SomaScan Menu?
Remember that the 7k menu contains <em>all</em> of the
<code>SeqIds</code> in the 5k menu, so what this really tells us is:
were analytes targeting cell adhesion-related proteins added in the 7k
menu?</p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cellAd_menu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"5k"</span>, <span class="st">"7k"</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/select-SomaDb-method.html">select</a></span><span class="op">(</span><span class="va">SomaScan.db</span>, keys <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">cellAd_prots</span><span class="op">$</span><span class="va">UNIPROT</span><span class="op">)</span>,</span>
<span>                 keytype <span class="op">=</span> <span class="st">"UNIPROT"</span>, columns <span class="op">=</span> <span class="st">"PROBEID"</span>,</span>
<span>                 menu <span class="op">=</span> <span class="va">x</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="co"># Again, removing probes that do not map to a cell adhesion protein</span></span>
<span>    <span class="va">df</span> <span class="op">&lt;-</span> <span class="va">df</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">PROBEID</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="op">}</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"somascan_5k"</span>, <span class="st">"somascan_7k"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'select()' returned 1:many mapping between keys and columns</span></span>
<span><span class="co">## 'select()' returned 1:many mapping between keys and columns</span></span></code></pre>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/identical.html" class="external-link">identical</a></span><span class="op">(</span><span class="va">cellAd_menu</span><span class="op">$</span><span class="va">somascan_5k</span>, <span class="va">cellAd_menu</span><span class="op">$</span><span class="va">somascan_7k</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<p>In this example, the number of <code>SeqIds</code> associated with
cell adhesion does <em>not</em> differ between SomaScan menu versions
(the list of <code>SeqIds</code> is identical). The differences between
menu versions can be explored with the <code>menu</code> argument of
<code>select</code>, or via the <code>somascan_menu</code> data object
(this is explained in the Introductory Vignette).</p>
<hr>
</div>
<div class="section level3">
<h3 id="gene-families">Gene Families<a class="anchor" aria-label="anchor" href="#gene-families"></a>
</h3>
<p>A number of gene families are targeted by reagents in the SomaScan
assay. How can these be interrogated using <code>SomaScan.db</code>? Is
the package capable of searching for/within specific gene families? The
answer is yes, but<br>
a specific function does not exist for analyzing gene families as a
whole. Instead, by using features of <code>select</code> and
<code>keys</code>, <code>SomaScan.db</code> can be queried for common
features connecting gene families of interest - more specifically, the
<code>match=</code> argument of <code>select</code> and the
<code>pattern=</code> argument of <code>keys</code> can be used to
retrieve gene family members that contain a common pattern in their
name.</p>
<p>The <code>keys</code> method is capable of using regular expressions
(“regex”) to search for keys in the database that contain a specific
pattern of characters. This feature is especially useful when looking
for annotations for a gene family. For example, a regex pattern can be
used to retrieve a list of all IL17 receptor family genes in the
database:</p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">il17_family</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/keys-SomaDb-method.html">keys</a></span><span class="op">(</span><span class="va">SomaScan.db</span>, keytype <span class="op">=</span> <span class="st">"SYMBOL"</span>, pattern <span class="op">=</span> <span class="st">"IL17"</span><span class="op">)</span></span></code></pre></div>
<p>Those keys can then be used to query the database with
<code>select</code>:</p>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/select-SomaDb-method.html">select</a></span><span class="op">(</span><span class="va">SomaScan.db</span>, keys <span class="op">=</span> <span class="va">il17_family</span>, keytype <span class="op">=</span> <span class="st">"SYMBOL"</span>,</span>
<span>       columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PROBEID"</span>, <span class="st">"UNIPROT"</span>, <span class="st">"GENENAME"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'select()' returned 1:many mapping between keys and columns</span></span></code></pre>
<pre><code><span><span class="co">##     SYMBOL  PROBEID UNIPROT                  GENENAME</span></span>
<span><span class="co">## 1    IL17A  21897-4  Q16552           interleukin 17A</span></span>
<span><span class="co">## 2    IL17A  3498-53  Q16552           interleukin 17A</span></span>
<span><span class="co">## 3    IL17A  9170-24  Q16552           interleukin 17A</span></span>
<span><span class="co">## 4   IL17RA  2992-59  Q96F46 interleukin 17 receptor A</span></span>
<span><span class="co">## 5    IL17C   9255-5  Q9P0M4           interleukin 17C</span></span>
<span><span class="co">## 6    IL17B 14022-17  Q9UHF5           interleukin 17B</span></span>
<span><span class="co">## 7    IL17B  3499-77  Q9UHF5           interleukin 17B</span></span>
<span><span class="co">## 8    IL17D  4136-40  Q8TAD2           interleukin 17D</span></span>
<span><span class="co">## 9   IL17RD  3376-49  Q8NFM7 interleukin 17 receptor D</span></span>
<span><span class="co">## 10  IL17RD  3376-49  B4DXM5 interleukin 17 receptor D</span></span>
<span><span class="co">## 11  IL17RB 5084-154  Q9NRM6 interleukin 17 receptor B</span></span>
<span><span class="co">## 12  IL17RB  6262-14  Q9NRM6 interleukin 17 receptor B</span></span>
<span><span class="co">## 13  IL17RC  5468-67  Q8NAC3 interleukin 17 receptor C</span></span>
<span><span class="co">## 14   IL17F 14026-24  Q96PD4           interleukin 17F</span></span>
<span><span class="co">## 15   IL17F  21897-4  Q96PD4           interleukin 17F</span></span>
<span><span class="co">## 16   IL17F  2775-54  Q96PD4           interleukin 17F</span></span>
<span><span class="co">## 17  IL17RE 20535-68  B4DMZ3 interleukin 17 receptor E</span></span>
<span><span class="co">## 18  IL17RE 20535-68  Q8NFR9 interleukin 17 receptor E</span></span>
<span><span class="co">## 19 IL17REL     &lt;NA&gt;    &lt;NA&gt;                      &lt;NA&gt;</span></span></code></pre>
<p>If multiple gene families are of interest, the <code>keys</code>
argument of <code>select</code> (in combination with
<code>match=TRUE</code>) can support a regex pattern, and will
accomplish both of the previous steps in a single call:</p>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/select-SomaDb-method.html">select</a></span><span class="op">(</span><span class="va">SomaScan.db</span>, keys <span class="op">=</span> <span class="st">"NOTCH|ZF"</span>, keytype <span class="op">=</span> <span class="st">"SYMBOL"</span>, </span>
<span>       columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PROBEID"</span>, <span class="st">"SYMBOL"</span>, <span class="st">"GENENAME"</span><span class="op">)</span>, match <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'select()' returned 1:many mapping between keys and columns</span></span></code></pre>
<pre><code><span><span class="co">##     SYMBOL   PROBEID</span></span>
<span><span class="co">## 1   NOTCH2  11297-54</span></span>
<span><span class="co">## 2   NOTCH2   5106-52</span></span>
<span><span class="co">## 3   NOTCH2   8407-84</span></span>
<span><span class="co">## 4  ZFYVE27   13432-9</span></span>
<span><span class="co">## 5  ZFYVE27   9102-28</span></span>
<span><span class="co">## 6    ZFP91  13651-54</span></span>
<span><span class="co">## 7     MZF1   14662-6</span></span>
<span><span class="co">## 8   ZFAND5 18317-111</span></span>
<span><span class="co">## 9   ZFAND1   19173-5</span></span>
<span><span class="co">## 10  CREBZF   21134-9</span></span>
<span><span class="co">## 11  ZFAND3  21875-31</span></span>
<span><span class="co">## 12   ZFP36   22395-7</span></span>
<span><span class="co">## 13 ZFAND2B   23319-6</span></span>
<span><span class="co">## 14  NOTCH1    5107-7</span></span>
<span><span class="co">## 15  NOTCH3   5108-72</span></span>
<span><span class="co">##                                                   GENENAME</span></span>
<span><span class="co">## 1                                         notch receptor 2</span></span>
<span><span class="co">## 2                                         notch receptor 2</span></span>
<span><span class="co">## 3                                         notch receptor 2</span></span>
<span><span class="co">## 4                      zinc finger FYVE-type containing 27</span></span>
<span><span class="co">## 5                      zinc finger FYVE-type containing 27</span></span>
<span><span class="co">## 6  ZFP91 zinc finger protein, atypical E3 ubiquitin ligase</span></span>
<span><span class="co">## 7                                    myeloid zinc finger 1</span></span>
<span><span class="co">## 8                        zinc finger AN1-type containing 5</span></span>
<span><span class="co">## 9                        zinc finger AN1-type containing 1</span></span>
<span><span class="co">## 10                      CREB/ATF bZIP transcription factor</span></span>
<span><span class="co">## 11                       zinc finger AN1-type containing 3</span></span>
<span><span class="co">## 12                               ZFP36 ring finger protein</span></span>
<span><span class="co">## 13                      zinc finger AN1-type containing 2B</span></span>
<span><span class="co">## 14                                        notch receptor 1</span></span>
<span><span class="co">## 15                                        notch receptor 3</span></span></code></pre>
<p>The <code>GENENAME</code> column can also support a regex pattern,
and can be used to search for keywords that are associated with specific
gene families (and not just the gene symbols themselves). Examples
include “homeobox”, “zinc finger”, “notch”, etc.</p>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/select-SomaDb-method.html">select</a></span><span class="op">(</span><span class="va">SomaScan.db</span>, keys <span class="op">=</span> <span class="st">"homeobox"</span>, keytype <span class="op">=</span> <span class="st">"GENENAME"</span>, </span>
<span>       columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PROBEID"</span>, <span class="st">"SYMBOL"</span><span class="op">)</span>, match <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'select()' returned 1:1 mapping between keys and columns</span></span></code></pre>
<pre><code><span><span class="co">##       GENENAME   PROBEID SYMBOL</span></span>
<span><span class="co">## 1 homeobox A11  22375-15 HOXA11</span></span>
<span><span class="co">## 2  homeobox A5  22376-95  HOXA5</span></span>
<span><span class="co">## 3 homeobox C11  22474-28 HOXC11</span></span>
<span><span class="co">## 4  homeobox D4 22476-115  HOXD4</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.2 (2023-10-31)</span></span>
<span><span class="co">## Platform: x86_64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">## Running under: macOS Monterey 12.6.9</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] withr_2.5.2          SomaScan.db_0.99.8   org.Hs.eg.db_3.18.0 </span></span>
<span><span class="co">##  [4] KEGGREST_1.42.0      GO.db_3.18.0         AnnotationDbi_1.64.1</span></span>
<span><span class="co">##  [7] IRanges_2.36.0       S4Vectors_0.40.2     Biobase_2.62.0      </span></span>
<span><span class="co">## [10] BiocGenerics_0.48.1  BiocStyle_2.30.0    </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] sass_0.4.8              bitops_1.0-7           </span></span>
<span><span class="co">##  [3] RSQLite_2.3.4           stringi_1.8.3          </span></span>
<span><span class="co">##  [5] digest_0.6.33           magrittr_2.0.3         </span></span>
<span><span class="co">##  [7] evaluate_0.23           bookdown_0.37          </span></span>
<span><span class="co">##  [9] fastmap_1.1.1           blob_1.2.4             </span></span>
<span><span class="co">## [11] jsonlite_1.8.8          GenomeInfoDb_1.38.2    </span></span>
<span><span class="co">## [13] DBI_1.1.3               BiocManager_1.30.22    </span></span>
<span><span class="co">## [15] httr_1.4.7              purrr_1.0.2            </span></span>
<span><span class="co">## [17] Biostrings_2.70.1       textshaping_0.3.7      </span></span>
<span><span class="co">## [19] jquerylib_0.1.4         cli_3.6.2              </span></span>
<span><span class="co">## [21] rlang_1.1.2             crayon_1.5.2           </span></span>
<span><span class="co">## [23] XVector_0.42.0          bit64_4.0.5            </span></span>
<span><span class="co">## [25] cachem_1.0.8            yaml_2.3.8             </span></span>
<span><span class="co">## [27] tools_4.3.2             memoise_2.0.1          </span></span>
<span><span class="co">## [29] GenomeInfoDbData_1.2.11 curl_5.2.0             </span></span>
<span><span class="co">## [31] vctrs_0.6.5             R6_2.5.1               </span></span>
<span><span class="co">## [33] png_0.1-8               lifecycle_1.0.4        </span></span>
<span><span class="co">## [35] zlibbioc_1.48.0         stringr_1.5.1          </span></span>
<span><span class="co">## [37] fs_1.6.3                bit_4.0.5              </span></span>
<span><span class="co">## [39] ragg_1.2.7              pkgconfig_2.0.3        </span></span>
<span><span class="co">## [41] desc_1.4.3              pkgdown_2.0.7          </span></span>
<span><span class="co">## [43] bslib_0.6.1             glue_1.6.2             </span></span>
<span><span class="co">## [45] systemfonts_1.0.5       xfun_0.41              </span></span>
<span><span class="co">## [47] knitr_1.45              htmltools_0.5.7        </span></span>
<span><span class="co">## [49] rmarkdown_2.25          compiler_4.3.2         </span></span>
<span><span class="co">## [51] RCurl_1.98-1.13</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



   </div>
  <footer><div class="container">
  <div class="pkgdown-footer-left">
  <p></p>
<p>SomaScan.db was developed by the Bioinformatics Dept. at <a href="http://www.somalogic.com" class="external-link">SomaLogic Operating Co., Inc.</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

  </div></footer>
</body>
</html>
