<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="SomaScan.db">
<title>Introduction to SomaScan.db • SomaScan.db</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Roboto-0.4.9/font.css" rel="stylesheet">
<link href="../deps/Source_Code_Pro-0.4.9/font.css" rel="stylesheet">
<!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to SomaScan.db">
<meta property="og:description" content="SomaScan.db">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#container" class="visually-hidden-focusable">Skip to content</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-none" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">SomaScan.db</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.99.10</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/SomaScan-db.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/SomaScan-db.html">Introduction to SomaScan.db</a>
    <a class="dropdown-item" href="../articles/example_adat_workflow.html">Example ADAT Workflow Using SomaScan.db</a>
    <a class="dropdown-item" href="../articles/advanced_usage_examples.html">Advanced Exploration of the SomaScan Menu</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/SomaLogic/SomaScan.db/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article" id="container">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Introduction to SomaScan.db</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/SomaLogic/SomaScan.db/blob/HEAD/vignettes/SomaScan-db.Rmd" class="external-link"><code>vignettes/SomaScan-db.Rmd</code></a></small>
      <div class="d-none name"><code>SomaScan-db.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The <code>SomaScan.db</code> package provides extended biological
annotations to be used in conjunction with the results of SomaLogic’s
SomaScan assay, a fee-for-service proteomic technology platform designed
to detect proteins across numerous biological pathways.</p>
<p>This vignette describes how to use the <code>SomaScan.db</code>
package to annotate SomaScan data, i.e. add additional information to an
ADAT file that will give biological context (at the gene level) to the
platform’s reagents and their protein targets. <code>SomaScan.db</code>
performs annotation by mapping SomaScan reagent IDs
(<code>SeqIds</code>) to their corresponding protein(s) and gene(s), as
well as biological pathways (GO, KEGG, etc.) and identifiers from other
public data repositories.</p>
<p><code>SomaScan.db</code> utilizes the same methods and setup as other
Bioconductor annotation packages, and therefore the methods should be
familiar if you’ve worked with such packages previously.</p>
</div>
<div class="section level2">
<h2 id="package-installation">Package Installation<a class="anchor" aria-label="anchor" href="#package-installation"></a>
</h2>
<p>To begin, install and load the <code>SomaScan.db</code> package:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"SomaScan.db"</span>, version <span class="op">=</span> <span class="fu">remotes</span><span class="fu">::</span><span class="fu">bioc_version</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Once installed, the package can be loaded as follows:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://somalogic.com" class="external-link">SomaScan.db</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="package-overview">Package Overview<a class="anchor" aria-label="anchor" href="#package-overview"></a>
</h2>
<p>Loading this package will expose an annotation object with the same
name as the package, <code>SomaScan.db</code>. This object is a <a href="https://www.sqlite.org/index.html" class="external-link">SQLite database</a> containing
annotation data for the SomaScan assay derived from popular public
repositories. Viewing the object will present a metadata table
containing information about the annotations and where they were
obtained:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SomaScan.db</span></span></code></pre></div>
<pre><code><span><span class="co">## SomaDb object:</span></span>
<span><span class="co">## | DBSCHEMAVERSION: 2.1</span></span>
<span><span class="co">## | Db type: ChipDb</span></span>
<span><span class="co">## | Supporting package: SomaScan.db</span></span>
<span><span class="co">## | DBSCHEMA: HUMANCHIP_DB</span></span>
<span><span class="co">## | ORGANISM: Homo sapiens</span></span>
<span><span class="co">## | SPECIES: Human</span></span>
<span><span class="co">## | MANUFACTURER: SomaLogic</span></span>
<span><span class="co">## | CHIPNAME: SomaScan</span></span>
<span><span class="co">## | MANUFACTURERURL: https://somalogic.com/somascan-platform/</span></span>
<span><span class="co">## | EGSOURCEDATE: 2023-Sep11</span></span>
<span><span class="co">## | EGSOURCENAME: Entrez Gene</span></span>
<span><span class="co">## | EGSOURCEURL: ftp://ftp.ncbi.nlm.nih.gov/gene/DATA</span></span>
<span><span class="co">## | CENTRALID: ENTREZID</span></span>
<span><span class="co">## | TAXID: 9606</span></span>
<span><span class="co">## | GOSOURCENAME: Gene Ontology</span></span>
<span><span class="co">## | GOSOURCEURL: http://current.geneontology.org/ontology/go-basic.obo</span></span>
<span><span class="co">## | GOSOURCEDATE: 2023-07-27</span></span>
<span><span class="co">## | GOEGSOURCEDATE: 2023-Sep11</span></span>
<span><span class="co">## | GOEGSOURCENAME: Entrez Gene</span></span>
<span><span class="co">## | GOEGSOURCEURL: ftp://ftp.ncbi.nlm.nih.gov/gene/DATA</span></span>
<span><span class="co">## | KEGGSOURCENAME: KEGG GENOME</span></span>
<span><span class="co">## | KEGGSOURCEURL: ftp://ftp.genome.jp/pub/kegg/genomes</span></span>
<span><span class="co">## | KEGGSOURCEDATE: 2011-Mar15</span></span>
<span><span class="co">## | GPSOURCENAME: UCSC Genome Bioinformatics (Homo sapiens)</span></span>
<span><span class="co">## | GPSOURCEURL: </span></span>
<span><span class="co">## | GPSOURCEDATE: 2023-Aug20</span></span>
<span><span class="co">## | ENSOURCEDATE: 2023-May10</span></span>
<span><span class="co">## | ENSOURCENAME: Ensembl</span></span>
<span><span class="co">## | ENSOURCEURL: ftp://ftp.ensembl.org/pub/current_fasta</span></span>
<span><span class="co">## | UPSOURCENAME: Uniprot</span></span>
<span><span class="co">## | UPSOURCEURL: http://www.UniProt.org/</span></span>
<span><span class="co">## | UPSOURCEDATE: Mon Sep 18 16:12:39 2023</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Please see: help('select') for usage information</span></span></code></pre>
<p>The same information can be retrieved as a data frame by calling
<code>metadata(SomaScan.db)</code>.</p>
<p>Moving forward, this database object (<code>SomaScan.db</code>) will
be used throughout the vignette to retrieve SomaScan annotations and map
between identifiers.</p>
<p>For reference, the species information for the database can directly
be retrieved with the following methods:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/organism_species.html" class="external-link">species</a></span><span class="op">(</span><span class="va">SomaScan.db</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Homo sapiens"</span></span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">taxonomyId</a></span><span class="op">(</span><span class="va">SomaScan.db</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 9606</span></span></code></pre>
<p>It’s also possible to pull a more detailed summary of annotations and
resource identifiers (aka keys) by calling the package as a function,
with the <code>.db</code> extension removed:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/SomaScanBASE.html">SomaScan</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Quality control information for SomaScan:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## This package has the following mappings:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## SomaScanALIAS2PROBE has 38839 mapped keys (of 258270 keys)</span></span>
<span><span class="co">## SomaScanENSEMBL has 10596 mapped keys (of 10731 keys)</span></span>
<span><span class="co">## SomaScanENSEMBL2PROBE has 11012 mapped keys (of 40467 keys)</span></span>
<span><span class="co">## SomaScanENTREZID has 10609 mapped keys (of 10731 keys)</span></span>
<span><span class="co">## SomaScanENZYME has 1712 mapped keys (of 10731 keys)</span></span>
<span><span class="co">## SomaScanENZYME2PROBE has 809 mapped keys (of 975 keys)</span></span>
<span><span class="co">## SomaScanGENENAME has 10609 mapped keys (of 10731 keys)</span></span>
<span><span class="co">## SomaScanGO has 10499 mapped keys (of 10731 keys)</span></span>
<span><span class="co">## SomaScanGO2ALLPROBES has 20327 mapped keys (of 22741 keys)</span></span>
<span><span class="co">## SomaScanGO2PROBE has 16066 mapped keys (of 18809 keys)</span></span>
<span><span class="co">## SomaScanMAP has 10607 mapped keys (of 10731 keys)</span></span>
<span><span class="co">## SomaScanOMIM has 9766 mapped keys (of 10731 keys)</span></span>
<span><span class="co">## SomaScanPATH has 4099 mapped keys (of 10731 keys)</span></span>
<span><span class="co">## SomaScanPATH2PROBE has 229 mapped keys (of 229 keys)</span></span>
<span><span class="co">## SomaScanPMID has 10607 mapped keys (of 10731 keys)</span></span>
<span><span class="co">## SomaScanPMID2PROBE has 605433 mapped keys (of 778807 keys)</span></span>
<span><span class="co">## SomaScanREFSEQ has 10605 mapped keys (of 10731 keys)</span></span>
<span><span class="co">## SomaScanSYMBOL has 10609 mapped keys (of 10731 keys)</span></span>
<span><span class="co">## SomaScanUNIPROT has 10545 mapped keys (of 10731 keys)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Additional Information about this package:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## DB schema: HUMANCHIP_DB</span></span>
<span><span class="co">## DB schema version: 2.1</span></span>
<span><span class="co">## Organism: Homo sapiens</span></span>
<span><span class="co">## Date for NCBI data: 2023-Sep11</span></span>
<span><span class="co">## Date for GO data: 2023-07-27</span></span>
<span><span class="co">## Date for KEGG data: 2011-Mar15</span></span>
<span><span class="co">## Date for Golden Path data: 2023-Aug20</span></span>
<span><span class="co">## Date for Ensembl data: 2023-May10</span></span></code></pre>
<p><em>Note</em>: Keys will be explained in greater detail later in this
vignette.</p>
</div>
<div class="section level2">
<h2 id="retrieve-annotation-data">Retrieve Annotation Data<a class="anchor" aria-label="anchor" href="#retrieve-annotation-data"></a>
</h2>
<p>The <code>SomaScan.db</code> package has 5 primary methods that can
be used to query the database:</p>
<ul>
<li><code>keys</code></li>
<li><code>keytypes</code></li>
<li><code>columns</code></li>
<li><code>select</code></li>
<li><code>mapIds</code></li>
</ul>
<p>This vignette will describe how each of these methods can be used to
obtain annotation data from <code>SomaScan.db</code>.</p>
<hr>
<div class="section level3">
<h3 id="keys-method">
<code>keys</code> method<a class="anchor" aria-label="anchor" href="#keys-method"></a>
</h3>
<p>This annotation package is platform-based, meaning it was built
around the unique identifiers from a specific platform (in this case,
SomaLogic’s SomaScan platform). That identifier corresponds to each of
the assay’s analytes, and therefore the analyte identifiers
(<code>SeqIds</code>) are the primary term used to query the database
(aka “key”).</p>
<p>All keys in the database can be retrieved using
<code>keys</code>:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Short list of primary keys</span></span>
<span><span class="fu"><a href="../reference/keys-SomaDb-method.html">keys</a></span><span class="op">(</span><span class="va">SomaScan.db</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">10L</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "10000-28"  "10001-7"   "10003-15"  "10006-25"  "10008-43" </span></span>
<span><span class="co">##  [6] "10010-10"  "10011-65"  "10012-5"   "10014-31"  "10015-119"</span></span></code></pre>
<p>Each key retrieved in the output above corresponds to one of the
assay’s unique analytes.</p>
<hr>
</div>
<div class="section level3">
<h3 id="keytype-method">
<code>keytype</code> method<a class="anchor" aria-label="anchor" href="#keytype-method"></a>
</h3>
<p>When querying the database, we can also specify the type of key
(“keytype”) being used. The keytype refers to the type of identifier
that is used to generate a database query. While the database is
centered around the SomaLogic <code>SeqId</code>, other identifiers can
still be used to query the database.</p>
<p>We can list all available datatypes that can be used as query keys
using <code><a href="../reference/keytypes-SomaDb-method.html">keytypes()</a></code>:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## List all of the supported key types.</span></span>
<span><span class="fu"><a href="../reference/keytypes-SomaDb-method.html">keytypes</a></span><span class="op">(</span><span class="va">SomaScan.db</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "ACCNUM"       "ALIAS"        "ENSEMBL"      "ENSEMBLPROT" </span></span>
<span><span class="co">##  [5] "ENSEMBLTRANS" "ENTREZID"     "ENZYME"       "EVIDENCE"    </span></span>
<span><span class="co">##  [9] "EVIDENCEALL"  "GENENAME"     "GENETYPE"     "GO"          </span></span>
<span><span class="co">## [13] "GOALL"        "IPI"          "MAP"          "OMIM"        </span></span>
<span><span class="co">## [17] "ONTOLOGY"     "ONTOLOGYALL"  "PATH"         "PFAM"        </span></span>
<span><span class="co">## [21] "PMID"         "PROBEID"      "PROSITE"      "REFSEQ"      </span></span>
<span><span class="co">## [25] "SYMBOL"       "UCSCKG"       "UNIPROT"</span></span></code></pre>
<p><em>Note:</em> the SomaScan assay analyte identifiers
(<code>SeqIds</code>) are stored as the “PROBEID” keytype.</p>
<p><code>keytypes</code> can also be used in conjunction with
<code>keys</code> to retrieve all identifiers associated with the
specified keytype. The example below will retrieve all UniProt IDs in
<code>SomaScan.db</code>:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/keys-SomaDb-method.html">keys</a></span><span class="op">(</span><span class="va">SomaScan.db</span>, keytype <span class="op">=</span> <span class="st">"UNIPROT"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">20L</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "P04217"     "V9HWD8"     "P01023"     "P18440"     "Q400J6"    </span></span>
<span><span class="co">##  [6] "F5H5R8"     "A4Z6T7"     "P11245"     "A0A024R6P0" "P01011"    </span></span>
<span><span class="co">## [11] "P22760"     "A0A024R410" "Q13685"     "C9JEH3"     "F1T0I5"    </span></span>
<span><span class="co">## [16] "Q16613"     "P49588"     "P80404"     "X5D8S1"     "B2RUU2"</span></span></code></pre>
<hr>
</div>
<div class="section level3">
<h3 id="columns-method">
<code>columns</code> method<a class="anchor" aria-label="anchor" href="#columns-method"></a>
</h3>
<p>All available external annotations, corresponding to “columns” of the
database, can be listed using <code><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">columns()</a></code>:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">columns</a></span><span class="op">(</span><span class="va">SomaScan.db</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "ACCNUM"       "ALIAS"        "ENSEMBL"      "ENSEMBLPROT" </span></span>
<span><span class="co">##  [5] "ENSEMBLTRANS" "ENTREZID"     "ENZYME"       "EVIDENCE"    </span></span>
<span><span class="co">##  [9] "EVIDENCEALL"  "GENENAME"     "GENETYPE"     "GO"          </span></span>
<span><span class="co">## [13] "GOALL"        "IPI"          "MAP"          "OMIM"        </span></span>
<span><span class="co">## [17] "ONTOLOGY"     "ONTOLOGYALL"  "PATH"         "PFAM"        </span></span>
<span><span class="co">## [21] "PMID"         "PROBEID"      "PROSITE"      "REFSEQ"      </span></span>
<span><span class="co">## [25] "SYMBOL"       "UCSCKG"       "UNIPROT"</span></span></code></pre>
<p><em>Note:</em> the SomaScan assay analyte identifiers
(<code>SeqIds</code>) are stored in the “PROBEID” column.</p>
<p>This list may look very similar (or even identical) to the
<code>columns</code> output. If identical, all columns can be used as
query keys. For a more in-depth explanation of what each of these
columns contains, consult the manual:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/help.html" class="external-link">help</a></span><span class="op">(</span><span class="st">"OMIM"</span><span class="op">)</span> <span class="co"># Example help call</span></span></code></pre></div>
<p>Each <code>columns</code> entry also has a mapping object that
contains the information connecting SeqIds → the annotation’s
identifiers. To read further documentation about the object and the
resource used to make it, check out the manual page for the mapping
itself:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">?</span><span class="va">SomaScanOMIM</span></span></code></pre></div>
<hr>
</div>
<div class="section level3">
<h3 id="select-method">
<code>select</code> method<a class="anchor" aria-label="anchor" href="#select-method"></a>
</h3>
<p>The list of columns returned by <code>columns</code> informs us as to
what types of data are available; therefore, the column values can be
used to retrieve specific pieces of information from the database. You
can think of keys and columns as:</p>
<ul>
<li>
<strong>Keys</strong>: the information you already have
(<code>SeqIds</code>/probe IDs), aka <em>rows</em> of the database</li>
<li>
<strong>Columns</strong>: data types for which you want to retrieve
information, aka <em>columns</em> of the database</li>
</ul>
<p>The <code>SomaScan.db</code> database can be queried via
<code>select</code>, using both the keys and columns.</p>
<p>When selecting columns and keys using the <code>select</code> method,
the keys are returned in the left-most column of the output, in the
<code>PROBEID</code> column. The results will be in the exact same order
as the input keys:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Randomly select a set of keys</span></span>
<span><span class="va">example_keys</span> <span class="op">&lt;-</span> <span class="fu">withr</span><span class="fu">::</span><span class="fu"><a href="https://withr.r-lib.org/reference/with_seed.html" class="external-link">with_seed</a></span><span class="op">(</span><span class="fl">101L</span>, <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-envirAPI.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="../reference/keys-SomaDb-method.html">keys</a></span><span class="op">(</span><span class="va">SomaScan.db</span><span class="op">)</span>,</span>
<span>                                              size <span class="op">=</span> <span class="fl">5L</span>,</span>
<span>                                              replace <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Query keys in the database</span></span>
<span><span class="fu"><a href="../reference/select-SomaDb-method.html">select</a></span><span class="op">(</span><span class="va">SomaScan.db</span>,</span>
<span>       keys <span class="op">=</span> <span class="va">example_keys</span>,</span>
<span>       columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ENTREZID"</span>, <span class="st">"SYMBOL"</span>, <span class="st">"GENENAME"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'select()' returned 1:1 mapping between keys and columns</span></span></code></pre>
<pre><code><span><span class="co">##    PROBEID ENTREZID  SYMBOL</span></span>
<span><span class="co">## 1 34799-31    81563 C1orf21</span></span>
<span><span class="co">## 2 20520-14    10447   FAM3C</span></span>
<span><span class="co">## 3  9123-18   255057   CBARP</span></span>
<span><span class="co">## 4 21707-15   389941   C1QL3</span></span>
<span><span class="co">## 5 29325-43    11102   RPP14</span></span>
<span><span class="co">##                                          GENENAME</span></span>
<span><span class="co">## 1              chromosome 1 open reading frame 21</span></span>
<span><span class="co">## 2 FAM3 metabolism regulating signaling molecule C</span></span>
<span><span class="co">## 3 CACN subunit beta associated regulatory protein</span></span>
<span><span class="co">## 4                           complement C1q like 3</span></span>
<span><span class="co">## 5                  ribonuclease P/MRP subunit p14</span></span></code></pre>
<p><strong>Note:</strong> The message above (<strong>‘select()’ returned
1:1 mapping between keys and columns</strong>) will be described in
detail in the next section of this vignette.</p>
<p>The data that is returned will always be in the same order as the
provided keys. If <code>select</code> cannot find a mapping for a
specific key, an <code>NA</code> value will be returned to retain the
original query order.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Inserting a new key that won't be found in the annotations ("TEST")</span></span>
<span><span class="va">test_keys</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">example_keys</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"TEST"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/select-SomaDb-method.html">select</a></span><span class="op">(</span><span class="va">SomaScan.db</span>, keys <span class="op">=</span> <span class="va">test_keys</span>, columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PROBEID"</span>, <span class="st">"ENTREZID"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'select()' returned 1:1 mapping between keys and columns</span></span></code></pre>
<pre><code><span><span class="co">##    PROBEID ENTREZID</span></span>
<span><span class="co">## 1 34799-31    81563</span></span>
<span><span class="co">## 2     &lt;NA&gt;     &lt;NA&gt;</span></span></code></pre>
<p>In the example above, a “PROBEID” and “ENTREZID” value couldn’t be
found for the character string “TEST”, so an <code>NA</code> was
returned in its place.</p>
<div class="section level4">
<h4 id="one-to-many-relationships">One-to-Many Relationships<a class="anchor" aria-label="anchor" href="#one-to-many-relationships"></a>
</h4>
<p>When using <code>select</code>, a message indicating the relationship
between query keys and column data will be displayed along with the
query results. This message will describe one of the three relationships
below:</p>
<ul>
<li><code>1:1 mapping between keys and columns</code></li>
<li><code>1:many mapping between keys and columns</code></li>
<li><code>many:many mapping between keys and columns</code></li>
</ul>
<p>These messages describe the very real possibility that there are
multiple identifiers associated with each key in a query. This can cause
the number of rows returned by <code><a href="../reference/select-SomaDb-method.html">select()</a></code> to exceed the
number of keys used to retrieve the data; this is what is meant by the
message “‘select()’ returned 1:many mapping between keys and
columns”.</p>
<p>In these cases, you will still see concordance between the order of
the provided keys and outputted results rows, but you should be aware
that new rows were inserted into the results. This message is not an
error, merely an informative notification to the user making it clear
that more output rows than input items should be expected. Importantly,
this message also does not relay information about the SomaScan menu
itself or advice on how to handle many-to-one relationships between
SomaScan reagents and their corresponding protein targets; rather, the
message is directly related to this package’s <code>select</code> method
and how it retrieves information from the database.</p>
<p>Because some columns may have a many-to-many relationship to each
key, it is generally best practice to retrieve the <strong>minimum
number of columns needed</strong> for a query. Additionally, when
retrieving a column that is <em>known</em> to have a many-to-one
relationship to each key, like GO terms, it’s best to request that
information in its own query, like so:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Good</span></span>
<span><span class="fu"><a href="../reference/select-SomaDb-method.html">select</a></span><span class="op">(</span><span class="va">SomaScan.db</span>, keys <span class="op">=</span> <span class="va">example_keys</span><span class="op">[</span><span class="fl">3L</span><span class="op">]</span>, columns <span class="op">=</span> <span class="st">"GO"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'select()' returned 1:many mapping between keys and columns</span></span></code></pre>
<pre><code><span><span class="co">##    PROBEID         GO EVIDENCE ONTOLOGY</span></span>
<span><span class="co">## 1  9123-18 GO:0005886      IBA       CC</span></span>
<span><span class="co">## 2  9123-18 GO:0005886      ISS       CC</span></span>
<span><span class="co">## 3  9123-18 GO:0030141      IBA       CC</span></span>
<span><span class="co">## 4  9123-18 GO:0030141      ISS       CC</span></span>
<span><span class="co">## 5  9123-18 GO:0030426      ISS       CC</span></span>
<span><span class="co">## 6  9123-18 GO:0030672      IEA       CC</span></span>
<span><span class="co">## 7  9123-18 GO:0044325      IBA       MF</span></span>
<span><span class="co">## 8  9123-18 GO:0044325      ISS       MF</span></span>
<span><span class="co">## 9  9123-18 GO:0045955      IBA       BP</span></span>
<span><span class="co">## 10 9123-18 GO:0045955      ISS       BP</span></span>
<span><span class="co">## 11 9123-18 GO:1901386      IBA       BP</span></span>
<span><span class="co">## 12 9123-18 GO:1901386      ISS       BP</span></span>
<span><span class="co">## 13 9123-18 GO:1903170      ISS       BP</span></span></code></pre>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Bad</span></span>
<span><span class="fu"><a href="../reference/select-SomaDb-method.html">select</a></span><span class="op">(</span><span class="va">SomaScan.db</span>,</span>
<span>       keys <span class="op">=</span> <span class="va">example_keys</span><span class="op">[</span><span class="fl">3L</span><span class="op">]</span>,</span>
<span>       columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"UNIPROT"</span>, <span class="st">"ENSEMBL"</span>, <span class="st">"GO"</span>, <span class="st">"PATH"</span>, <span class="st">"IPI"</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: You have selected the following columns that can have a many to</span></span>
<span><span class="co">##   one relationship with the primary key: UNIPROT, ENSEMBL, GO,</span></span>
<span><span class="co">##   PATH, IPI . Because you have selected more than a few such</span></span>
<span><span class="co">##   columns there is a risk that this selection may balloon up into</span></span>
<span><span class="co">##   a very large result as the number of rows returned multiplies</span></span>
<span><span class="co">##   accordingly. To experience smaller/more manageable results and</span></span>
<span><span class="co">##   faster retrieval times, you might want to consider selecting</span></span>
<span><span class="co">##   these columns separately.</span></span></code></pre>
<pre><code><span><span class="co">## 'select()' returned 1:many mapping between keys and columns</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 26 × 8</span></span></span>
<span><span class="co">##    PROBEID UNIPROT ENSEMBL         GO        EVIDENCE ONTOLOGY PATH  IPI  </span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> 9123-18 Q8N350  ENSG00000099625 GO:00058… IBA      CC       <span style="color: #BB0000;">NA</span>    IPI0…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> 9123-18 Q8N350  ENSG00000099625 GO:00058… IBA      CC       <span style="color: #BB0000;">NA</span>    IPI0…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> 9123-18 Q8N350  ENSG00000099625 GO:00058… ISS      CC       <span style="color: #BB0000;">NA</span>    IPI0…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> 9123-18 Q8N350  ENSG00000099625 GO:00058… ISS      CC       <span style="color: #BB0000;">NA</span>    IPI0…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> 9123-18 Q8N350  ENSG00000099625 GO:00301… IBA      CC       <span style="color: #BB0000;">NA</span>    IPI0…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> 9123-18 Q8N350  ENSG00000099625 GO:00301… IBA      CC       <span style="color: #BB0000;">NA</span>    IPI0…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> 9123-18 Q8N350  ENSG00000099625 GO:00301… ISS      CC       <span style="color: #BB0000;">NA</span>    IPI0…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> 9123-18 Q8N350  ENSG00000099625 GO:00301… ISS      CC       <span style="color: #BB0000;">NA</span>    IPI0…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> 9123-18 Q8N350  ENSG00000099625 GO:00304… ISS      CC       <span style="color: #BB0000;">NA</span>    IPI0…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> 9123-18 Q8N350  ENSG00000099625 GO:00304… ISS      CC       <span style="color: #BB0000;">NA</span>    IPI0…</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 16 more rows</span></span></span></code></pre>
<p>The example above illustrates why it is preferred to request as few
columns as possible, especially when working with GO terms.</p>
</div>
<div class="section level4">
<h4 id="specifying-keytype">Specifying Keytype<a class="anchor" aria-label="anchor" href="#specifying-keytype"></a>
</h4>
<p>In <code>SomaScan.db</code>, the default keytype for
<code>select</code> is the probe ID. This means that when using a
<code>SeqId</code> (aka “PROBEID”) to retrieve annotations, the
<code>keytype=</code> argument does <em>not</em> need to be defined, and
can be left out of the <code>select</code> call entirely. The default
(<code>PROBEID</code>) will be used.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/select-SomaDb-method.html">select</a></span><span class="op">(</span><span class="va">SomaScan.db</span>, keys <span class="op">=</span> <span class="va">example_keys</span>, columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ENTREZID"</span>, <span class="st">"UNIPROT"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'select()' returned 1:many mapping between keys and columns</span></span></code></pre>
<pre><code><span><span class="co">##    PROBEID ENTREZID    UNIPROT</span></span>
<span><span class="co">## 1 34799-31    81563 A0A024R926</span></span>
<span><span class="co">## 2 34799-31    81563     Q9H246</span></span>
<span><span class="co">## 3 20520-14    10447     Q92520</span></span>
<span><span class="co">## 4  9123-18   255057     Q8N350</span></span>
<span><span class="co">## 5 21707-15   389941     Q5VWW1</span></span>
<span><span class="co">## 6 29325-43    11102     O95059</span></span></code></pre>
<p>However, the database can be searched using more than just
<code>SeqIds</code>.</p>
<p>For example, you may want to retrieve a list of <code>SeqIds</code>
that are associated with a specific gene of interest - let’s use SMAD2
as an example. You can work “backwards” to retrieve the
<code>SeqIds</code> associated with SMAD2 by setting the
<code>keytype="SYMBOL"</code>:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/select-SomaDb-method.html">select</a></span><span class="op">(</span><span class="va">SomaScan.db</span>,</span>
<span>       columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PROBEID"</span>, <span class="st">"ENTREZID"</span><span class="op">)</span>,</span>
<span>       keys <span class="op">=</span> <span class="st">"SMAD2"</span>, keytype <span class="op">=</span> <span class="st">"SYMBOL"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'select()' returned 1:many mapping between keys and columns</span></span></code></pre>
<pre><code><span><span class="co">##   SYMBOL   PROBEID ENTREZID</span></span>
<span><span class="co">## 1  SMAD2   10364-6     4087</span></span>
<span><span class="co">## 2  SMAD2 11353-143     4087</span></span></code></pre>
<p>Sometimes, this may <em>appear</em> not to work. Let’s use CASC4 as
an example:</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/select-SomaDb-method.html">select</a></span><span class="op">(</span><span class="va">SomaScan.db</span>,</span>
<span>       columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PROBEID"</span>, <span class="st">"ENTREZID"</span><span class="op">)</span>,</span>
<span>       keys <span class="op">=</span> <span class="st">"CASC4"</span>, keytype <span class="op">=</span> <span class="st">"SYMBOL"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Error in .testForValidKeys(x, keys, keytype, fks): None of the keys entered are valid keys for 'SYMBOL'. Please use the keys method to see a listing of valid arguments.</span></span></code></pre>
<p>The error message above implies that <code>CASC4</code> is not a
valid key for the “SYMBOL” keytype, which means that no entry for CASC4
was found in the “SYMBOL” column. However, genes can be tricky to
search, and in some cases have many common names. We can improve this
using <code>keytype="ALIAS"</code>; this data type contains the various
aliases associated with gene names found in the “SYMBOL” column. Using
<code>keytype="ALIAS"</code>, we can cast a wider net:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/select-SomaDb-method.html">select</a></span><span class="op">(</span><span class="va">SomaScan.db</span>,</span>
<span>       columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SYMBOL"</span>, <span class="st">"PROBEID"</span>, <span class="st">"ENTREZID"</span><span class="op">)</span>,</span>
<span>       keys <span class="op">=</span> <span class="st">"CASC4"</span>, keytype <span class="op">=</span> <span class="st">"ALIAS"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'select()' returned 1:many mapping between keys and columns</span></span></code></pre>
<pre><code><span><span class="co">##   ALIAS SYMBOL  PROBEID ENTREZID</span></span>
<span><span class="co">## 1 CASC4  GOLM2 10613-33   113201</span></span>
<span><span class="co">## 2 CASC4  GOLM2  8838-10   113201</span></span></code></pre>
<p>This reveals the source of our problem! CASC4 is also known as GOLM4,
and this symbol is used in the annotations database. Because of this,
searching for CASC4 as a symbol returns no results, but the same query
is able to identify an entry when the “ALIAS” column is specified.</p>
<p>Additionally, we can see that CASC4/GOLM2 is associated with two
<code>SeqIds</code> - <code>10613-33</code> and <code>8838-10</code>.
How is this possible, and why does this happen? For more information,
please reference the Advanced Usage Examples (
<code><a href="../articles/advanced_usage_examples.html">vignette("advanced_usage_examples", package = "SomaScan.db")</a></code>).</p>
</div>
<div class="section level4">
<h4 id="specifying-menu-version">Specifying Menu Version<a class="anchor" aria-label="anchor" href="#specifying-menu-version"></a>
</h4>
<p><code>SomaScan.db</code> contains annotations for all currently
available versions of the SomaScan menu by default. However, the
<code>menu=</code> argument of <code>select</code> can be used to
retrieve analytes from a single, specified menu. For example, if you
have SomaScan data from an older menu version, like the <code>5k</code>
menu (also known as <code>V4.0</code>), this argument can be used to
retrieve annotations associated with that menu specifically:</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">all_keys</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/keys-SomaDb-method.html">keys</a></span><span class="op">(</span><span class="va">SomaScan.db</span><span class="op">)</span></span>
<span><span class="va">menu_5k</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/select-SomaDb-method.html">select</a></span><span class="op">(</span><span class="va">SomaScan.db</span>,</span>
<span>                  keys <span class="op">=</span> <span class="va">all_keys</span>, columns <span class="op">=</span> <span class="st">"PROBEID"</span>,</span>
<span>                  menu <span class="op">=</span> <span class="st">"5k"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">menu_5k</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    PROBEID</span></span>
<span><span class="co">## 1 10000-28</span></span>
<span><span class="co">## 2  10001-7</span></span>
<span><span class="co">## 3 10003-15</span></span>
<span><span class="co">## 4 10006-25</span></span>
<span><span class="co">## 5 10008-43</span></span>
<span><span class="co">## 6 10011-65</span></span></code></pre>
<p>The <code>SeqIds</code> in the <code>menu_5k</code> data frame
represent the 4966 analytes that were available in v4.0 of the SomaScan
menu. All of the listed analytes have currently available annotations in
<code>SomaScan.db</code>. Those that are not represented do not
currently have annotations available in <code>SomaScan.db</code>.</p>
<p>If the <code>menu=</code> argument is not specified in
<code>select</code>, annotations for <em>all</em> available analytes are
returned.</p>
<hr>
</div>
</div>
<div class="section level3">
<h3 id="mapids-method">
<code>mapIDs</code> method<a class="anchor" aria-label="anchor" href="#mapids-method"></a>
</h3>
<p>For situations in which you wish only to retrieve one data type from
the database, the <code>mapIds</code> method may be cleaner and more
streamlined than using <code>select</code>, and can help avoid problems
with one-to-many mapping of keys. For example, if you are only
interested in the gene symbols associated with a set of SomaScan
analytes, they can be retrieved like so:</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/mapIds-SomaDb-method.html">mapIds</a></span><span class="op">(</span><span class="va">SomaScan.db</span>, keys <span class="op">=</span> <span class="va">example_keys</span>, column <span class="op">=</span> <span class="st">"SYMBOL"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'select()' returned 1:1 mapping between keys and columns</span></span></code></pre>
<pre><code><span><span class="co">##  34799-31  20520-14   9123-18  21707-15  29325-43 </span></span>
<span><span class="co">## "C1orf21"   "FAM3C"   "CBARP"   "C1QL3"   "RPP14"</span></span></code></pre>
<p><code>mapIds</code> will return a <strong>named vector</strong> from
a <em>single</em> column, while <code>select</code> returns a data frame
and can be used to retrieve data from multiple columns.</p>
<p>The primary difference between <code>mapIds</code> and
<code>select</code> is how the method handles one-to-many mapping,
i.e. when the chosen key maps to &gt; 1 entry in the selected column.
When this occurs, only the <strong>first value</strong> (by default) is
returned.</p>
<p>Compare the output in the examples below:</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Only 1 symbol per key</span></span>
<span><span class="fu"><a href="../reference/mapIds-SomaDb-method.html">mapIds</a></span><span class="op">(</span><span class="va">SomaScan.db</span>, keys <span class="op">=</span> <span class="va">example_keys</span><span class="op">[</span><span class="fl">3L</span><span class="op">]</span>, column <span class="op">=</span> <span class="st">"GO"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'select()' returned 1:many mapping between keys and columns</span></span></code></pre>
<pre><code><span><span class="co">##      9123-18 </span></span>
<span><span class="co">## "GO:0005886"</span></span></code></pre>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># All entries for chosen key</span></span>
<span><span class="fu"><a href="../reference/select-SomaDb-method.html">select</a></span><span class="op">(</span><span class="va">SomaScan.db</span>, keys <span class="op">=</span> <span class="va">example_keys</span><span class="op">[</span><span class="fl">3L</span><span class="op">]</span>, column <span class="op">=</span> <span class="st">"GO"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'select()' returned 1:many mapping between keys and columns</span></span></code></pre>
<pre><code><span><span class="co">##    PROBEID         GO EVIDENCE ONTOLOGY</span></span>
<span><span class="co">## 1  9123-18 GO:0005886      IBA       CC</span></span>
<span><span class="co">## 2  9123-18 GO:0005886      ISS       CC</span></span>
<span><span class="co">## 3  9123-18 GO:0030141      IBA       CC</span></span>
<span><span class="co">## 4  9123-18 GO:0030141      ISS       CC</span></span>
<span><span class="co">## 5  9123-18 GO:0030426      ISS       CC</span></span>
<span><span class="co">## 6  9123-18 GO:0030672      IEA       CC</span></span>
<span><span class="co">## 7  9123-18 GO:0044325      IBA       MF</span></span>
<span><span class="co">## 8  9123-18 GO:0044325      ISS       MF</span></span>
<span><span class="co">## 9  9123-18 GO:0045955      IBA       BP</span></span>
<span><span class="co">## 10 9123-18 GO:0045955      ISS       BP</span></span>
<span><span class="co">## 11 9123-18 GO:1901386      IBA       BP</span></span>
<span><span class="co">## 12 9123-18 GO:1901386      ISS       BP</span></span>
<span><span class="co">## 13 9123-18 GO:1903170      ISS       BP</span></span></code></pre>
<p>Note that the <code>mapIds</code> method warning message states that
it <em>returned</em> 1:many mappings between keys and columns, however
only one value was returned for the desired <code>SeqId</code>. This is
because there were more mapped values that were discarded when the
results were converted to a named vector. This may not be a problem for
some columns, like “SYMBOL” (typically there is only one gene symbol per
gene), but it may present a problem for others (like GO terms or KEGG
pathways). Think carefully when using <code>mapIds</code>, or consider
specifying the <code>multiVals=</code> argument to indicate what should
be done with multi-mapped output.</p>
<p>The default behavior of <code>mapIds</code> is to return the first
available result:</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># The default - returns the first available result</span></span>
<span><span class="fu"><a href="../reference/mapIds-SomaDb-method.html">mapIds</a></span><span class="op">(</span><span class="va">SomaScan.db</span>, keys <span class="op">=</span> <span class="va">example_keys</span><span class="op">[</span><span class="fl">3L</span><span class="op">]</span>, column <span class="op">=</span> <span class="st">"GO"</span>, </span>
<span>       multiVals <span class="op">=</span> <span class="st">"first"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'select()' returned 1:many mapping between keys and columns</span></span></code></pre>
<pre><code><span><span class="co">##      9123-18 </span></span>
<span><span class="co">## "GO:0005886"</span></span></code></pre>
<p>Again, the <code>select</code> message here indicates that, while
only 1 value was returned, there were many more GO term matches. All of
the matches can be viewed by specifying
<code>multiVals="list"</code>:</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Returns a list object of results, instead of only returning the first result</span></span>
<span><span class="fu"><a href="../reference/mapIds-SomaDb-method.html">mapIds</a></span><span class="op">(</span><span class="va">SomaScan.db</span>, keys <span class="op">=</span> <span class="va">example_keys</span><span class="op">[</span><span class="fl">3L</span><span class="op">]</span>, column <span class="op">=</span> <span class="st">"GO"</span>, </span>
<span>       multiVals <span class="op">=</span> <span class="st">"list"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'select()' returned 1:many mapping between keys and columns</span></span></code></pre>
<pre><code><span><span class="co">## $`9123-18`</span></span>
<span><span class="co">##  [1] "GO:0005886" "GO:0005886" "GO:0030141" "GO:0030141" "GO:0030426"</span></span>
<span><span class="co">##  [6] "GO:0030672" "GO:0044325" "GO:0044325" "GO:0045955" "GO:0045955"</span></span>
<span><span class="co">## [11] "GO:1901386" "GO:1901386" "GO:1903170"</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="adding-target-full-names">Adding Target Full Names<a class="anchor" aria-label="anchor" href="#adding-target-full-names"></a>
</h2>
<p>Because the annotations in this package are compiled from public
repositories, information typically found in an ADAT may be missing. For
example, in an ADAT file, each <code>SeqId</code> is associated with a
protein target, and the name of that target is provided as both an
abbreviated symbol (“Target”) and full description (“Target Full Name”).
The <code>SomaScan.db</code> package does not contain data <em>from</em>
a particular ADAT file; however, it does contain a function to add the
full protein target name to any data frame obtained via
<code>select</code>.</p>
<p>As an example, we will generate a data frame of <a href="https://useast.ensembl.org/info/about/index.html" class="external-link">Ensembl</a> gene
IDs and <a href="https://www.omim.org/" class="external-link">OMIM</a> IDs:</p>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ensg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/select-SomaDb-method.html">select</a></span><span class="op">(</span><span class="va">SomaScan.db</span>,</span>
<span>               keys <span class="op">=</span> <span class="va">example_keys</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3L</span><span class="op">]</span>,</span>
<span>               columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ENSEMBL"</span>, <span class="st">"OMIM"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'select()' returned 1:1 mapping between keys and columns</span></span></code></pre>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ensg</span></span></code></pre></div>
<pre><code><span><span class="co">##    PROBEID         ENSEMBL   OMIM</span></span>
<span><span class="co">## 1 34799-31 ENSG00000116667   &lt;NA&gt;</span></span>
<span><span class="co">## 2 20520-14 ENSG00000196937 608618</span></span>
<span><span class="co">## 3  9123-18 ENSG00000099625   &lt;NA&gt;</span></span></code></pre>
<p>We will now append the Target Full Name to this data frame:</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/addTargetFullName.html">addTargetFullName</a></span><span class="op">(</span><span class="va">ensg</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    PROBEID                  TARGETFULLNAME         ENSEMBL   OMIM</span></span>
<span><span class="co">## 1 20520-14                   Protein FAM3C ENSG00000196937 608618</span></span>
<span><span class="co">## 2 34799-31 Uncharacterized protein C1orf21 ENSG00000116667   &lt;NA&gt;</span></span>
<span><span class="co">## 3  9123-18                     Protein Dos ENSG00000099625   &lt;NA&gt;</span></span></code></pre>
<p>The full protein target name will be appended to the input data
frame, with the Target Full Name (in the “TARGETFULLNAME” column) always
added to the right of the “PROBEID” column.</p>
</div>
<div class="section level2">
<h2 id="package-objects">Package Objects<a class="anchor" aria-label="anchor" href="#package-objects"></a>
</h2>
<p>In addition to the methods mentioned above, there is an R object that
can be used to retrieve SomaScan analytes from a specific menu version.
The object is a list, with each element in the list containing a
character vector of <code>SeqIds</code> that were available in the
specified menu.</p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">somascan_menu</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      Length Class  Mode     </span></span>
<span><span class="co">## v4.0  4966  -none- character</span></span>
<span><span class="co">## v4.1  7267  -none- character</span></span>
<span><span class="co">## v5.0 10731  -none- character</span></span></code></pre>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">somascan_menu</span>, <span class="va">head</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $v4.0</span></span>
<span><span class="co">## [1] "10000-28" "10001-7"  "10003-15" "10006-25" "10008-43" "10011-65"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $v4.1</span></span>
<span><span class="co">## [1] "10000-28" "10001-7"  "10003-15" "10006-25" "10008-43" "10010-10"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $v5.0</span></span>
<span><span class="co">## [1] "10000-28" "10001-7"  "10003-15" "10006-25" "10008-43" "10010-10"</span></span></code></pre>
<p>This object also provides a quick and easy way of comparing the
available SomaScan menus:</p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">setdiff</a></span><span class="op">(</span><span class="va">somascan_menu</span><span class="op">$</span><span class="va">v4.1</span>, <span class="va">somascan_menu</span><span class="op">$</span><span class="va">v4.0</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">50L</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "10010-10"  "10025-1"   "10039-32"  "10069-2"   "10351-51" </span></span>
<span><span class="co">##  [6] "10354-57"  "10379-19"  "10382-1"   "10398-110" "10420-30" </span></span>
<span><span class="co">## [11] "10439-57"  "10457-3"   "10460-1"   "10463-23"  "10470-34" </span></span>
<span><span class="co">## [16] "10472-53"  "10473-2"   "10479-18"  "10480-33"  "10505-12" </span></span>
<span><span class="co">## [21] "10528-2"   "10576-7"   "10626-116" "10631-9"   "10636-1"  </span></span>
<span><span class="co">## [26] "10670-26"  "10738-11"  "10741-22"  "10743-13"  "10746-24" </span></span>
<span><span class="co">## [31] "10780-10"  "10801-11"  "10819-108" "10855-55"  "10870-32" </span></span>
<span><span class="co">## [36] "10894-25"  "10966-1"   "10967-12"  "10970-3"   "10976-44" </span></span>
<span><span class="co">## [41] "10980-11"  "11081-1"   "11083-23"  "11121-56"  "11150-3"  </span></span>
<span><span class="co">## [46] "11159-14"  "11184-51"  "11200-52"  "11203-97"  "11232-46"</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.3 (2024-02-29)</span></span>
<span><span class="co">## Platform: x86_64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">## Running under: macOS Monterey 12.7.4</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] tibble_3.2.1         SomaScan.db_0.99.10  AnnotationDbi_1.64.1</span></span>
<span><span class="co">## [4] IRanges_2.36.0       S4Vectors_0.40.2     Biobase_2.62.0      </span></span>
<span><span class="co">## [7] BiocGenerics_0.48.1  withr_3.0.0          BiocStyle_2.30.0    </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] utf8_1.2.4              sass_0.4.9             </span></span>
<span><span class="co">##  [3] bitops_1.0-7            RSQLite_2.3.6          </span></span>
<span><span class="co">##  [5] digest_0.6.35           magrittr_2.0.3         </span></span>
<span><span class="co">##  [7] evaluate_0.23           bookdown_0.39          </span></span>
<span><span class="co">##  [9] fastmap_1.1.1           blob_1.2.4             </span></span>
<span><span class="co">## [11] jsonlite_1.8.8          GenomeInfoDb_1.38.8    </span></span>
<span><span class="co">## [13] DBI_1.2.2               BiocManager_1.30.22    </span></span>
<span><span class="co">## [15] httr_1.4.7              fansi_1.0.6            </span></span>
<span><span class="co">## [17] purrr_1.0.2             Biostrings_2.70.3      </span></span>
<span><span class="co">## [19] textshaping_0.3.7       jquerylib_0.1.4        </span></span>
<span><span class="co">## [21] cli_3.6.2               rlang_1.1.3            </span></span>
<span><span class="co">## [23] crayon_1.5.2            XVector_0.42.0         </span></span>
<span><span class="co">## [25] bit64_4.0.5             cachem_1.0.8           </span></span>
<span><span class="co">## [27] yaml_2.3.8              tools_4.3.3            </span></span>
<span><span class="co">## [29] memoise_2.0.1           GenomeInfoDbData_1.2.11</span></span>
<span><span class="co">## [31] org.Hs.eg.db_3.18.0     vctrs_0.6.5            </span></span>
<span><span class="co">## [33] R6_2.5.1                png_0.1-8              </span></span>
<span><span class="co">## [35] lifecycle_1.0.4         zlibbioc_1.48.2        </span></span>
<span><span class="co">## [37] KEGGREST_1.42.0         fs_1.6.3               </span></span>
<span><span class="co">## [39] bit_4.0.5               ragg_1.3.0             </span></span>
<span><span class="co">## [41] pkgconfig_2.0.3         desc_1.4.3             </span></span>
<span><span class="co">## [43] pillar_1.9.0            pkgdown_2.0.9          </span></span>
<span><span class="co">## [45] bslib_0.7.0             glue_1.7.0             </span></span>
<span><span class="co">## [47] systemfonts_1.0.6       xfun_0.43              </span></span>
<span><span class="co">## [49] knitr_1.45              htmltools_0.5.8.1      </span></span>
<span><span class="co">## [51] rmarkdown_2.26          compiler_4.3.3         </span></span>
<span><span class="co">## [53] RCurl_1.98-1.14</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



   </div>
  <footer><div class="container">
  <div class="pkgdown-footer-left">
  <p>SomaScan.db was developed by the Bioinformatics Dept. at <a href="http://www.somalogic.com" class="external-link">SomaLogic Operating Co., Inc.</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

  </div></footer>
</body>
</html>
